// @mastra/core/dist/logger/index.cjs
'use strict';

var chunkQ5I3BBUU_cjs = require('../chunk-Q5I3BBUU.cjs');
var chunk4CVPHNP6_cjs = require('../chunk-4CVPHNP6.cjs');



Object.defineProperty(exports, "DualLogger", {
  enumerable: true,
  get: function () { return chunkQ5I3BBUU_cjs.DualLogger; }
});
Object.defineProperty(exports, "MultiLogger", {
  enumerable: true,
  get: function () { return chunkQ5I3BBUU_cjs.MultiLogger; }
});
Object.defineProperty(exports, "createLogger", {
  enumerable: true,
  get: function () { return chunkQ5I3BBUU_cjs.createLogger; }
});
Object.defineProperty(exports, "noopLogger", {
  enumerable: true,
  get: function () { return chunkQ5I3BBUU_cjs.noopLogger; }
});
Object.defineProperty(exports, "ConsoleLogger", {
  enumerable: true,
  get: function () { return chunk4CVPHNP6_cjs.ConsoleLogger; }
});
Object.defineProperty(exports, "LogLevel", {
  enumerable: true,
  get: function () { return chunk4CVPHNP6_cjs.LogLevel; }
});
Object.defineProperty(exports, "LoggerTransport", {
  enumerable: true,
  get: function () { return chunk4CVPHNP6_cjs.LoggerTransport; }
});
Object.defineProperty(exports, "MastraLogger", {
  enumerable: true,
  get: function () { return chunk4CVPHNP6_cjs.MastraLogger; }
});
Object.defineProperty(exports, "RegisteredLogger", {
  enumerable: true,
  get: function () { return chunk4CVPHNP6_cjs.RegisteredLogger; }
});
Object.defineProperty(exports, "createCustomTransport", {
  enumerable: true,
  get: function () { return chunk4CVPHNP6_cjs.createCustomTransport; }
});
