// @mastra/core/dist/auth/ee/index.cjs
'use strict';

var chunkIV2LFMTD_cjs = require('../../chunk-IV2LFMTD.cjs');
var chunkR5BRW53L_cjs = require('../../chunk-R5BRW53L.cjs');



Object.defineProperty(exports, "ACTIONS", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.ACTIONS; }
});
Object.defineProperty(exports, "DEFAULT_ROLES", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.DEFAULT_ROLES; }
});
Object.defineProperty(exports, "MastraFGAPermissions", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.MastraFGAPermissions; }
});
Object.defineProperty(exports, "PERMISSIONS", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.PERMISSIONS; }
});
Object.defineProperty(exports, "PERMISSION_PATTERNS", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.PERMISSION_PATTERNS; }
});
Object.defineProperty(exports, "RESOURCES", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.RESOURCES; }
});
Object.defineProperty(exports, "StaticRBACProvider", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.StaticRBACProvider; }
});
Object.defineProperty(exports, "buildCapabilities", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.buildCapabilities; }
});
Object.defineProperty(exports, "getDefaultRole", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.getDefaultRole; }
});
Object.defineProperty(exports, "hasPermission", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.hasPermission; }
});
Object.defineProperty(exports, "isAuthenticated", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.isAuthenticated; }
});
Object.defineProperty(exports, "isValidPermissionPattern", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.isValidPermissionPattern; }
});
Object.defineProperty(exports, "matchesPermission", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.matchesPermission; }
});
Object.defineProperty(exports, "resolvePermissions", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.resolvePermissions; }
});
Object.defineProperty(exports, "resolvePermissionsFromMapping", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.resolvePermissionsFromMapping; }
});
Object.defineProperty(exports, "validatePermissions", {
  enumerable: true,
  get: function () { return chunkIV2LFMTD_cjs.validatePermissions; }
});
Object.defineProperty(exports, "FGADeniedError", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.FGADeniedError; }
});
Object.defineProperty(exports, "checkFGA", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.checkFGA; }
});
Object.defineProperty(exports, "clearLicenseCache", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.clearLicenseCache; }
});
Object.defineProperty(exports, "getAgentFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getAgentFGAResourceId; }
});
Object.defineProperty(exports, "getAgentToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getAgentToolFGAResourceId; }
});
Object.defineProperty(exports, "getMCPToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getMCPToolFGAResourceId; }
});
Object.defineProperty(exports, "getStandaloneToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getStandaloneToolFGAResourceId; }
});
Object.defineProperty(exports, "getWorkflowFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getWorkflowFGAResourceId; }
});
Object.defineProperty(exports, "isDevEnvironment", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.isDevEnvironment; }
});
Object.defineProperty(exports, "isEEEnabled", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.isEEEnabled; }
});
Object.defineProperty(exports, "isEELicenseValid", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.isEELicenseValid; }
});
Object.defineProperty(exports, "isFeatureEnabled", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.isFeatureEnabled; }
});
Object.defineProperty(exports, "isLicenseValid", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.isLicenseValid; }
});
Object.defineProperty(exports, "requireFGA", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.requireFGA; }
});
Object.defineProperty(exports, "validateLicense", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.validateLicense; }
});
