// @mastra/core/dist/chunk-OVDUY2GP.js
import { createTool, isValidationError } from './chunk-LWKMPCC7.js';
import { isStandardSchemaWithJSON, standardSchemaToJSONSchema } from './chunk-6SRTDZ7S.js';
import { z } from 'zod/v4';
import { randomBytes } from 'crypto';
import { mkdtemp, writeFile, rm } from 'fs/promises';
import { tmpdir } from 'os';
import { join } from 'path';
import { pathToFileURL } from 'url';

// src/tools/code-mode/stub-generator.ts
var SAFE_IDENT = /^[A-Za-z_$][A-Za-z0-9_$]*$/;
function jsonSchemaToTsString(schema) {
  if (schema === void 0) return "unknown";
  if (typeof schema === "boolean") return schema ? "unknown" : "never";
  if (schema.const !== void 0) return literal(schema.const);
  if (Array.isArray(schema.enum)) {
    return schema.enum.length ? schema.enum.map(literal).join(" | ") : "never";
  }
  const union = schema.anyOf ?? schema.oneOf;
  if (Array.isArray(union) && union.length) {
    return union.map(jsonSchemaToTsString).join(" | ");
  }
  const type = normalizeType(schema.type);
  if (type === "object" || schema.properties) {
    return objectType(schema);
  }
  if (type === "array" || schema.items) {
    return arrayType(schema);
  }
  switch (type) {
    case "string":
      return "string";
    case "number":
    case "integer":
      return "number";
    case "boolean":
      return "boolean";
    case "null":
      return "null";
    default:
      return "unknown";
  }
}
function normalizeType(type) {
  if (Array.isArray(type)) {
    return type.find((t) => t !== "null");
  }
  return type;
}
function objectType(schema) {
  const props = schema.properties ?? {};
  const required = new Set(schema.required ?? []);
  const keys = Object.keys(props);
  if (!keys.length) {
    const additional = schema.additionalProperties;
    if (additional !== void 0 && additional !== false) {
      const valueType = typeof additional === "object" ? jsonSchemaToTsString(additional) : "unknown";
      return `Record<string, ${valueType}>`;
    }
    return "Record<string, unknown>";
  }
  const fields = keys.map((key) => {
    const optional = !required.has(key) ? "?" : "";
    const k = SAFE_IDENT.test(key) ? key : JSON.stringify(key);
    return `${k}${optional}: ${jsonSchemaToTsString(props[key])}`;
  });
  return `{ ${fields.join("; ")} }`;
}
function arrayType(schema) {
  const items = schema.items;
  if (Array.isArray(items)) {
    return `[${items.map(jsonSchemaToTsString).join(", ")}]`;
  }
  const inner = jsonSchemaToTsString(items);
  return isTopLevelUnion(inner) ? `Array<${inner}>` : `${inner}[]`;
}
function isTopLevelUnion(ts) {
  let depth = 0;
  for (let i = 0; i < ts.length; i++) {
    const c = ts[i];
    if (c === "{" || c === "[" || c === "(" || c === "<") depth++;
    else if (c === "}" || c === "]" || c === ")" || c === ">") depth--;
    else if (c === "|" && depth === 0) return true;
  }
  return false;
}
function literal(value) {
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number" || typeof value === "boolean") return String(value);
  if (value === null) return "null";
  return "unknown";
}
function schemaToTs(schema, io) {
  if (!isStandardSchemaWithJSON(schema)) return "unknown";
  try {
    const json = standardSchemaToJSONSchema(schema, { io });
    return jsonSchemaToTsString(json);
  } catch {
    return "unknown";
  }
}
function sanitizeId(id) {
  const cleaned = id.replace(/[^A-Za-z0-9_$]/g, "_");
  return SAFE_IDENT.test(cleaned) ? cleaned : `_${cleaned}`;
}
function generateStubs(tools) {
  const seen = /* @__PURE__ */ new Map();
  return Object.entries(tools).map(([key, tool]) => {
    const toolId = tool.id ?? key;
    const description = tool.description;
    const inputType = schemaToTs(tool.inputSchema, "input");
    const outputType = schemaToTs(tool.outputSchema, "output");
    const externalName = sanitizeId(toolId);
    const prior = seen.get(externalName);
    if (prior !== void 0 && prior !== toolId) {
      throw new Error(`Code Mode tool id collision: "${prior}" and "${toolId}" both map to external_${externalName}`);
    }
    seen.set(externalName, toolId);
    const doc = description ? `/** ${description.replace(/\*\//g, "* /")} */
` : "";
    const declaration = `${doc}declare function external_${externalName}(input: ${inputType}): Promise<${outputType}>;`;
    return { toolId, externalName, declaration };
  });
}
var USAGE_CONTRACT = `# Code Mode

You have an \`execute_typescript\` tool. Instead of calling tools one at a time,
write a single TypeScript program that orchestrates them and returns one result.

Rules:
- Call the available tools via the \`external_*\` functions declared below. Each
  returns a Promise \u2014 \`await\` it.
- Batch independent calls with \`Promise.all\`. Do arithmetic and data shaping in
  JavaScript, not in your head.
- End the program by \`return\`-ing the final value (objects/arrays are fine).
- The program runs in a sandbox: no access to the host filesystem, network, or
  process. The only capabilities are the \`external_*\` functions.
- Use \`console.log\` for debugging; logs are captured and returned.

Available functions:`;
function createCodeModeInstructions(config) {
  const stubs = generateStubs(config.tools);
  const declarations = stubs.map((s) => s.declaration).join("\n\n");
  return `${USAGE_CONTRACT}

${declarations}`;
}

// src/workspace/errors.ts
var WorkspaceError = class extends Error {
  constructor(message, code, workspaceId) {
    super(message);
    this.code = code;
    this.workspaceId = workspaceId;
    this.name = "WorkspaceError";
  }
  code;
  workspaceId;
};
var WorkspaceNotAvailableError = class extends WorkspaceError {
  constructor() {
    super("Workspace not available. Ensure the agent has a workspace configured.", "NO_WORKSPACE");
    this.name = "WorkspaceNotAvailableError";
  }
};
var FilesystemNotAvailableError = class extends WorkspaceError {
  constructor() {
    super("Workspace does not have a filesystem configured", "NO_FILESYSTEM");
    this.name = "FilesystemNotAvailableError";
  }
};
var SandboxNotAvailableError = class extends WorkspaceError {
  constructor(message) {
    super(message ?? "Workspace does not have a sandbox configured", "NO_SANDBOX");
    this.name = "SandboxNotAvailableError";
  }
};
var SandboxFeatureNotSupportedError = class extends WorkspaceError {
  constructor(feature) {
    super(`Sandbox does not support ${feature}`, "FEATURE_NOT_SUPPORTED");
    this.name = "SandboxFeatureNotSupportedError";
  }
};
var SearchNotAvailableError = class extends WorkspaceError {
  constructor() {
    super("Workspace does not have search configured (enable bm25 or provide vectorStore + embedder)", "NO_SEARCH");
    this.name = "SearchNotAvailableError";
  }
};
var WorkspaceNotReadyError = class extends WorkspaceError {
  constructor(workspaceId, status) {
    super(`Workspace is not ready (status: ${status})`, "NOT_READY", workspaceId);
    this.name = "WorkspaceNotReadyError";
  }
};
var WorkspaceReadOnlyError = class extends WorkspaceError {
  constructor(operation) {
    super(`Workspace is in read-only mode. Cannot perform: ${operation}`, "READ_ONLY");
    this.name = "WorkspaceReadOnlyError";
  }
};
var FilesystemError = class extends Error {
  constructor(message, code, path) {
    super(message);
    this.code = code;
    this.path = path;
    this.name = "FilesystemError";
  }
  code;
  path;
};
var FileNotFoundError = class extends FilesystemError {
  constructor(path) {
    super(`File not found: ${path}`, "ENOENT", path);
    this.name = "FileNotFoundError";
  }
};
var DirectoryNotFoundError = class extends FilesystemError {
  constructor(path) {
    super(`Directory not found: ${path}`, "ENOENT", path);
    this.name = "DirectoryNotFoundError";
  }
};
var FileExistsError = class extends FilesystemError {
  constructor(path) {
    super(`File already exists: ${path}`, "EEXIST", path);
    this.name = "FileExistsError";
  }
};
var IsDirectoryError = class extends FilesystemError {
  constructor(path) {
    super(`Path is a directory: ${path}`, "EISDIR", path);
    this.name = "IsDirectoryError";
  }
};
var NotDirectoryError = class extends FilesystemError {
  constructor(path) {
    super(`Path is not a directory: ${path}`, "ENOTDIR", path);
    this.name = "NotDirectoryError";
  }
};
var DirectoryNotEmptyError = class extends FilesystemError {
  constructor(path) {
    super(`Directory not empty: ${path}`, "ENOTEMPTY", path);
    this.name = "DirectoryNotEmptyError";
  }
};
var PermissionError = class extends FilesystemError {
  constructor(path, operation) {
    super(`Permission denied: ${operation} on ${path}`, "EACCES", path);
    this.operation = operation;
    this.name = "PermissionError";
  }
  operation;
};
var FileReadRequiredError = class extends FilesystemError {
  constructor(path, reason) {
    super(reason, "EREAD_REQUIRED", path);
    this.name = "FileReadRequiredError";
  }
};
var StaleFileError = class extends FilesystemError {
  constructor(path, expectedMtime, actualMtime) {
    super(
      `File was modified externally: ${path} (expected mtime ${expectedMtime.toISOString()}, actual ${actualMtime.toISOString()})`,
      "ESTALE",
      path
    );
    this.expectedMtime = expectedMtime;
    this.actualMtime = actualMtime;
    this.name = "StaleFileError";
  }
  expectedMtime;
  actualMtime;
};
var FilesystemNotReadyError = class extends FilesystemError {
  constructor(id) {
    super(`Filesystem "${id}" is not ready. Call init() first or use ensureReady().`, "ENOTREADY", id);
    this.name = "FilesystemNotReadyError";
  }
};

// src/tools/code-mode/runner.ts
var FRAME_PREFIX = "\0CODEMODE\0";
function buildProgramModule(program) {
  return `export default async function () {
${program}
}
`;
}
function buildRunner({ programModule, externals }) {
  const SAFE_IDENT2 = /^[A-Za-z_$][A-Za-z0-9_$]*$/;
  const seen = /* @__PURE__ */ new Map();
  for (const { externalName, toolId } of externals) {
    if (!SAFE_IDENT2.test(externalName)) {
      throw new Error(`Invalid Code Mode external identifier: ${externalName}`);
    }
    const existing = seen.get(externalName);
    if (existing) {
      throw new Error(
        `Code Mode external identifier collision: tools "${existing}" and "${toolId}" both map to external_${externalName}`
      );
    }
    seen.set(externalName, toolId);
  }
  const externalsJson = JSON.stringify(externals.map(({ externalName, toolId }) => ({ externalName, toolId })));
  return `'use strict';
const FRAME_PREFIX = ${JSON.stringify(FRAME_PREFIX)};

function __emit(frame) {
  process.stdout.write(FRAME_PREFIX + JSON.stringify(frame) + '\\n');
}

// ---- console capture -------------------------------------------------------
for (const level of ['log', 'info', 'warn', 'error']) {
  console[level] = (...args) => {
    const message = args
      .map((a) => (typeof a === 'string' ? a : safeStringify(a)))
      .join(' ');
    __emit({ type: 'log', level, message });
  };
}
function safeStringify(value) {
  try { return JSON.stringify(value); } catch { return String(value); }
}

// ---- RPC bridge ------------------------------------------------------------
let __nextId = 0;
const __pending = new Map();

function __rpc(tool, args) {
  const id = __nextId++;
  return new Promise((resolve, reject) => {
    __pending.set(id, { resolve, reject });
    __emit({ type: 'rpc', id, tool, args });
  });
}

let __stdinBuffer = '';
process.stdin.setEncoding('utf8');
process.stdin.on('data', (chunk) => {
  __stdinBuffer += chunk;
  let idx;
  while ((idx = __stdinBuffer.indexOf('\\n')) >= 0) {
    const line = __stdinBuffer.slice(0, idx);
    __stdinBuffer = __stdinBuffer.slice(idx + 1);
    if (!line) continue;
    let frame;
    try { frame = JSON.parse(line); } catch { continue; }
    if (frame && frame.type === 'rpc-result') {
      const entry = __pending.get(frame.id);
      if (!entry) continue;
      __pending.delete(frame.id);
      if (frame.ok) entry.resolve(frame.result);
      else {
        const err = new Error(frame.error?.message || 'external tool failed');
        if (frame.error?.name) err.name = frame.error.name;
        entry.reject(err);
      }
    }
  }
});

// ---- externals -------------------------------------------------------------
for (const { externalName, toolId } of ${externalsJson}) {
  globalThis['external_' + externalName] = (input) => __rpc(toolId, input);
}

// ---- user program ----------------------------------------------------------
// The program lives in a sibling .ts module exporting a default async function;
// node strips its TypeScript types natively on import.
async function __main() {
  const mod = await import(${JSON.stringify(programModule)});
  return await mod.default();
}

__main()
  .then((result) => {
    __emit({ type: 'done', ok: true, result });
    process.exit(0);
  })
  .catch((error) => {
    __emit({
      type: 'done',
      ok: false,
      error: { message: error?.message ?? String(error), name: error?.name },
    });
    process.exit(0);
  });
`;
}

// src/tools/code-mode/transport.ts
var StdioCodeModeTransport = class {
  async run(opts) {
    const { sandbox, program, toolIds, dispatch, timeout, abortSignal, onExternalCall, onExternalResult } = opts;
    if (!sandbox.processes) {
      throw new SandboxFeatureNotSupportedError("processes");
    }
    const externals = toolIds.map((toolId) => ({ toolId, externalName: sanitize(toolId) }));
    const allowList = new Set(toolIds);
    const dir = await mkdtemp(join(tmpdir(), "mastra-code-mode-"));
    const suffix = randomBytes(4).toString("hex");
    const programPath = join(dir, `program-${suffix}.ts`);
    await writeFile(programPath, buildProgramModule(program), "utf8");
    const runnerSource = buildRunner({ programModule: pathToFileURL(programPath).href, externals });
    const runnerPath = join(dir, `runner-${suffix}.mjs`);
    await writeFile(runnerPath, runnerSource, "utf8");
    const logs = [];
    let done;
    let stdoutBuffer = "";
    let resolveDone;
    const donePromise = new Promise((resolve) => {
      resolveDone = resolve;
    });
    try {
      let handleFrame2 = function(frame) {
        switch (frame.type) {
          case "log":
            logs.push(frame.message);
            return;
          case "done":
            done = frame.ok ? { success: true, result: frame.result, logs } : { success: false, error: frame.error, logs };
            resolveDone();
            return;
          case "rpc":
            void serveRpc(frame.id, frame.tool, frame.args).catch(() => {
            });
            return;
        }
      }, notifyCall2 = function(tool, args) {
        try {
          onExternalCall?.(tool, args);
        } catch {
        }
      }, notifyResult2 = function(tool, durationMs, error) {
        try {
          onExternalResult?.(tool, durationMs, error);
        } catch {
        }
      };
      const handle = await sandbox.processes.spawn(`node --experimental-strip-types ${runnerPath}`, {
        cwd: dir,
        abortSignal,
        onStdout: (chunk) => {
          stdoutBuffer += chunk;
          let idx;
          while ((idx = stdoutBuffer.indexOf("\n")) >= 0) {
            const line = stdoutBuffer.slice(0, idx);
            stdoutBuffer = stdoutBuffer.slice(idx + 1);
            if (!line.startsWith(FRAME_PREFIX)) continue;
            let frame;
            try {
              frame = JSON.parse(line.slice(FRAME_PREFIX.length));
            } catch {
              continue;
            }
            handleFrame2(frame);
          }
        }
      });
      async function serveRpc(id, tool, args) {
        const started = Date.now();
        notifyCall2(tool, args);
        if (!allowList.has(tool)) {
          notifyResult2(tool, Date.now() - started, new Error("not allowed"));
          await respond(id, false, void 0, {
            message: `Tool "${tool}" is not available in Code Mode`,
            name: "NotAllowedError"
          });
          return;
        }
        try {
          const result = await dispatch(tool, args);
          notifyResult2(tool, Date.now() - started);
          await respond(id, true, result);
        } catch (error) {
          notifyResult2(tool, Date.now() - started, error);
          await respond(id, false, void 0, {
            message: error?.message ?? String(error),
            name: error?.name
          });
        }
      }
      async function respond(id, ok, result, error) {
        await handle.sendStdin(JSON.stringify({ type: "rpc-result", id, ok, result, error }) + "\n");
      }
      let timer;
      const timeoutPromise = new Promise((resolve) => {
        timer = setTimeout(() => resolve("timeout"), timeout);
      });
      const exitPromise = handle.wait().then(() => "exited");
      const outcome = await Promise.race([
        donePromise.then(() => "done"),
        exitPromise.catch(() => "exited"),
        timeoutPromise
      ]);
      if (timer) clearTimeout(timer);
      if (outcome === "timeout") {
        await handle.kill().catch(() => {
        });
        return {
          success: false,
          logs,
          error: { message: `Code Mode execution timed out after ${timeout}ms`, name: "TimeoutError" }
        };
      }
      if (!done) {
        await exitPromise.catch(() => {
        });
      }
      return done ?? {
        success: false,
        logs,
        error: { message: "Program exited without returning a result", name: "NoResultError" }
      };
    } finally {
      await rm(dir, { recursive: true, force: true }).catch(() => {
      });
    }
  }
};
function sanitize(id) {
  const cleaned = id.replace(/[^A-Za-z0-9_$]/g, "_");
  return /^[A-Za-z_$]/.test(cleaned) ? cleaned : `_${cleaned}`;
}

// src/tools/code-mode/code-mode.ts
var DEFAULT_TIMEOUT = 3e4;
var DEFAULT_TOOL_NAME = "execute_typescript";
var codeModeInputSchema = z.object({
  code: z.string().describe(
    "A TypeScript program that orchestrates the available external_* tools and returns a final value. Use Promise.all to batch calls; do arithmetic in JS. End with `return <value>`."
  )
});
var codeModeOutputSchema = z.object({
  success: z.boolean(),
  result: z.unknown().optional(),
  logs: z.array(z.string()).optional(),
  error: z.object({
    message: z.string(),
    name: z.string().optional(),
    line: z.number().optional()
  }).optional()
});
function indexToolsById(config) {
  const map = /* @__PURE__ */ new Map();
  for (const [key, tool] of Object.entries(config.tools)) {
    const id = tool.id ?? key;
    map.set(id, tool);
  }
  return map;
}
function createCodeModeTool(config, transport = new StdioCodeModeTransport()) {
  const timeout = config.timeout ?? DEFAULT_TIMEOUT;
  const id = config.id ?? DEFAULT_TOOL_NAME;
  const toolsById = indexToolsById(config);
  const toolIds = [...toolsById.keys()];
  return createTool({
    id,
    description: "Execute a TypeScript program that orchestrates the available tools in a sandbox. Prefer this over calling tools one at a time when a task needs multiple tool calls, batching, aggregation, or arithmetic.",
    inputSchema: codeModeInputSchema,
    outputSchema: codeModeOutputSchema,
    execute: async ({ code }, ctx) => {
      const sandbox = config.sandbox ?? ctx?.workspace?.sandbox;
      if (!sandbox) {
        throw new Error(
          "Code Mode requires a sandbox to run model-authored code, but none was configured. Pass one to createCodeMode({ tools, sandbox }), or run the agent in a workspace that provides a sandbox. To execute on the host (host privileges \u2014 only for trusted/local use), pass `sandbox: new LocalSandbox()`."
        );
      }
      const dispatch = async (toolId, args) => {
        const tool = toolsById.get(toolId);
        if (!tool?.execute) {
          throw new Error(`Tool "${toolId}" is not available in Code Mode`);
        }
        const result = await tool.execute(args, {
          mastra: ctx?.mastra,
          requestContext: ctx?.requestContext,
          abortSignal: ctx?.abortSignal,
          workspace: ctx?.workspace
        });
        if (isValidationError(result)) {
          throw new Error(result.message ?? `Invalid input for tool "${toolId}"`);
        }
        return result;
      };
      return ctx.observe.span(
        `code-mode:${id}`,
        () => transport.run({
          sandbox,
          program: code,
          toolIds,
          dispatch,
          timeout,
          abortSignal: ctx?.abortSignal,
          onExternalCall: (tool, args) => ctx.observe.log("info", "code-mode external call", { tool, args }),
          onExternalResult: (tool, durationMs, error) => ctx.observe.log(error ? "error" : "info", "code-mode external result", { tool, durationMs })
        })
      );
    }
  });
}
function createCodeMode(config, transport) {
  return {
    tool: createCodeModeTool(config, transport),
    instructions: createCodeModeInstructions(config)
  };
}

export { DirectoryNotEmptyError, DirectoryNotFoundError, FRAME_PREFIX, FileExistsError, FileNotFoundError, FileReadRequiredError, FilesystemError, FilesystemNotAvailableError, FilesystemNotReadyError, IsDirectoryError, NotDirectoryError, PermissionError, SandboxFeatureNotSupportedError, SandboxNotAvailableError, SearchNotAvailableError, StaleFileError, StdioCodeModeTransport, WorkspaceError, WorkspaceNotAvailableError, WorkspaceNotReadyError, WorkspaceReadOnlyError, buildProgramModule, buildRunner, createCodeMode, createCodeModeInstructions, createCodeModeTool, generateStubs, jsonSchemaToTsString };
