// @mastra/core/dist/fga-check-IMD7W3WT.cjs
'use strict';

var chunkR5BRW53L_cjs = require('./chunk-R5BRW53L.cjs');



Object.defineProperty(exports, "FGADeniedError", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.FGADeniedError; }
});
Object.defineProperty(exports, "checkFGA", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.checkFGA; }
});
Object.defineProperty(exports, "getAgentFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getAgentFGAResourceId; }
});
Object.defineProperty(exports, "getAgentToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getAgentToolFGAResourceId; }
});
Object.defineProperty(exports, "getMCPToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getMCPToolFGAResourceId; }
});
Object.defineProperty(exports, "getStandaloneToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getStandaloneToolFGAResourceId; }
});
Object.defineProperty(exports, "getWorkflowFGAResourceId", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.getWorkflowFGAResourceId; }
});
Object.defineProperty(exports, "requireFGA", {
  enumerable: true,
  get: function () { return chunkR5BRW53L_cjs.requireFGA; }
});
