// @mastra/core/dist/provider-registry-TI2S24TW.cjs
'use strict';

var chunkXKOQ7YF6_cjs = require('./chunk-XKOQ7YF6.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isOfflineMode", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.isOfflineMode; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.isValidModelId; }
});
Object.defineProperty(exports, "modelSupportsAttachments", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.modelSupportsAttachments; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.parseModelString; }
});
