// @mastra/core/dist/harness/v1/index.js
import { RequestContext } from '../../chunk-CW4HICJV.js';
import { randomUUID, createHash } from 'crypto';

var HARNESS_EVENT_ID_PREFIX = "harness-v1";
var HarnessValidationError = class extends Error {
  constructor(path, message) {
    super(`${path}: ${message}`);
    this.path = path;
    this.name = "HarnessValidationError";
  }
  path;
};
var HarnessEventSerializationError = class extends HarnessValidationError {
  constructor(eventType, reason, sessionId) {
    super("event.payload", `custom event "${eventType}" payload is not JSON-serializable: ${reason}`);
    this.eventType = eventType;
    this.reason = reason;
    this.sessionId = sessionId;
    this.name = "HarnessEventSerializationError";
  }
  eventType;
  reason;
  sessionId;
};
function formatHarnessEventId(epoch, sequence) {
  if (epoch.length === 0 || epoch.includes(":")) {
    throw new HarnessValidationError("eventId.epoch", 'epoch must be non-empty and must not contain ":"');
  }
  if (!Number.isSafeInteger(sequence) || sequence < 0) {
    throw new HarnessValidationError("eventId.sequence", "sequence must be a non-negative safe integer");
  }
  return `${HARNESS_EVENT_ID_PREFIX}:${epoch}:${sequence}`;
}
function parseHarnessEventId(eventId) {
  const parts = eventId.split(":");
  if (parts.length !== 3 || parts[0] !== HARNESS_EVENT_ID_PREFIX || parts[1] === "" || parts[2] === "") {
    throw new HarnessValidationError("lastEventId", "expected event id grammar harness-v1:<epoch>:<seq>");
  }
  const sequenceText = parts[2];
  if (!/^(0|[1-9][0-9]*)$/.test(sequenceText)) {
    throw new HarnessValidationError("lastEventId", "event id sequence must be an unsigned decimal integer");
  }
  const sequence = Number(sequenceText);
  if (!Number.isSafeInteger(sequence)) {
    throw new HarnessValidationError("lastEventId", "event id sequence must be within JavaScript safe integer range");
  }
  return { epoch: parts[1], sequence };
}
function sessionCreatedPayload(record) {
  return {
    resourceId: record.resourceId,
    threadId: record.threadId,
    parentSessionId: record.parentSessionId,
    modeId: record.modeId,
    modelId: record.modelId
  };
}
function assertCustomEventType(eventType) {
  if (RESERVED_EVENT_TYPES.has(eventType) || RESERVED_EVENT_PREFIXES.some((prefix) => eventType.startsWith(prefix))) {
    throw new HarnessValidationError("event.type", `"${eventType}" is reserved by the harness`);
  }
  if (!eventType.includes(".")) {
    throw new HarnessValidationError("event.type", "custom event types must be namespaced with a dot");
  }
}
function assertJsonSerializable(eventType, sessionId, value) {
  visitJsonValue(eventType, sessionId, value, /* @__PURE__ */ new WeakSet());
}
var EventEmitter = class _EventEmitter {
  listeners = [];
  epoch;
  seq;
  scope;
  onEvent;
  constructor(scope = {}, opts = {}) {
    this.scope = scope;
    this.onEvent = opts.onEvent;
    this.epoch = opts.epoch;
    this.seq = opts.nextSequence ?? 0;
    formatHarnessEventId(this.epoch ?? "pending-epoch", this.seq);
  }
  scoped(scope) {
    return new _EventEmitter(
      { ...this.scope, ...scope },
      {
        onEvent: (event) => this.forward(event)
      }
    );
  }
  subscribe(listener) {
    this.listeners.push(listener);
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index !== -1) this.listeners.splice(index, 1);
    };
  }
  emit(event, overrides) {
    const sessionId = overrides?.sessionId ?? this.scope.sessionId;
    this.validateCustomEvent(event, sessionId);
    const stamped = {
      ...event,
      id: formatHarnessEventId(this.epochId, this.seq++),
      timestamp: Date.now(),
      ...sessionId !== void 0 && { sessionId }
    };
    this.dispatch(stamped);
    return stamped;
  }
  forward(event) {
    this.dispatch(event);
  }
  get listenerCount() {
    return this.listeners.length;
  }
  get epochId() {
    this.epoch ??= randomUUID();
    return this.epoch;
  }
  validateCustomEvent(event, sessionId) {
    const eventType = event.type;
    if (typeof eventType !== "string" || RESERVED_EVENT_TYPES.has(eventType)) return;
    assertCustomEventType(eventType);
    if (Object.prototype.hasOwnProperty.call(event, "payload")) {
      assertJsonSerializable(eventType, sessionId, event.payload);
    }
  }
  dispatch(event) {
    if (this.onEvent) {
      try {
        const result = this.onEvent(event);
        if (result && typeof result.catch === "function") {
          result.catch((err) => console.error("[harness/v1] event persistence rejected:", err));
        }
      } catch (err) {
        console.error("[harness/v1] event persistence threw:", err);
      }
    }
    for (const listener of [...this.listeners]) {
      try {
        const result = listener(event);
        if (result && typeof result.catch === "function") {
          result.catch((err) => console.error("[harness/v1] event listener rejected:", err));
        }
      } catch (err) {
        console.error("[harness/v1] event listener threw:", err);
      }
    }
  }
};
var RESERVED_EVENT_TYPES = /* @__PURE__ */ new Set(["session_created", "mode_changed", "model_changed", "thread_cloned"]);
var RESERVED_EVENT_PREFIXES = ["session_", "thread_"];
function visitJsonValue(eventType, sessionId, value, seen) {
  if (value === void 0) throwSerialization(eventType, sessionId, "undefined");
  if (typeof value === "function") throwSerialization(eventType, sessionId, "function");
  if (typeof value === "symbol") throwSerialization(eventType, sessionId, "symbol");
  if (typeof value === "bigint") throwSerialization(eventType, sessionId, "bigint");
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean") return;
  if (value instanceof Date) throwSerialization(eventType, sessionId, "date");
  if (value instanceof Map) throwSerialization(eventType, sessionId, "map");
  if (value instanceof Set) throwSerialization(eventType, sessionId, "set");
  if (ArrayBuffer.isView(value)) throwSerialization(eventType, sessionId, "typed-array");
  if (typeof value !== "object") return;
  if (seen.has(value)) throwSerialization(eventType, sessionId, "cycle");
  seen.add(value);
  if (Array.isArray(value)) {
    for (const item of value) visitJsonValue(eventType, sessionId, item, seen);
    seen.delete(value);
    return;
  }
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) {
    throwSerialization(eventType, sessionId, "non-plain-object");
  }
  for (const item of Object.values(value)) {
    visitJsonValue(eventType, sessionId, item, seen);
  }
  seen.delete(value);
}
function throwSerialization(eventType, sessionId, reason) {
  throw new HarnessEventSerializationError(eventType, reason, sessionId);
}
var Session = class _Session {
  /** Stable identity. Frozen at construction. */
  #id;
  #ownerId;
  #resourceId;
  #threadId;
  #createdAt;
  #lastActivityAt;
  #memory;
  #events;
  // readonly parentSessionId?: string;
  // readonly subagentDepth: number;
  #modelId;
  #mode;
  constructor(config) {
    this.#id = config.id;
    this.#ownerId = config.ownerId;
    this.#resourceId = config.resourceId;
    this.#threadId = config.threadId;
    this.#mode = config.mode;
    this.#modelId = config.model;
    this.#createdAt = config.createdAt;
    this.#lastActivityAt = config.lastActivityAt;
    this.#memory = config.memory;
    this.#events = config.events;
  }
  get id() {
    return this.#id;
  }
  get ownerId() {
    return this.#ownerId;
  }
  get resourceId() {
    return this.#resourceId;
  }
  get threadId() {
    return this.#threadId;
  }
  get createdAt() {
    return this.#createdAt;
  }
  async clone(opts = {}) {
    const result = await (await this.#resolveMemory()).cloneThread({
      sourceThreadId: this.#threadId,
      newThreadId: opts.threadId,
      resourceId: opts.resourceId ?? this.#resourceId,
      title: opts.title,
      metadata: opts.metadata,
      options: opts.messageLimit !== void 0 ? { messageLimit: opts.messageLimit } : void 0
    });
    const cloneId = opts.sessionId ?? randomUUID();
    const clone = new _Session({
      id: cloneId,
      ownerId: this.#ownerId,
      threadId: result.thread.id,
      resourceId: result.thread.resourceId,
      mode: opts.mode ?? this.#mode,
      model: opts.modelId ?? this.#modelId,
      createdAt: result.thread.createdAt,
      lastActivityAt: result.thread.updatedAt,
      memory: this.#memory,
      events: this.#events.scoped({ sessionId: cloneId })
    });
    this.#events.emit({
      type: "thread_cloned",
      threadId: clone.threadId,
      resourceId: clone.resourceId,
      sourceThreadId: this.#threadId,
      title: opts.title
    });
    return clone;
  }
  async getThread() {
    return (await this.#resolveMemory()).getThreadById({ threadId: this.#threadId });
  }
  async getMessages() {
    const result = await (await this.#resolveMemory()).recall({ threadId: this.#threadId, resourceId: this.#resourceId });
    return result.messages;
  }
  async saveMessages(messages) {
    return (await this.#resolveMemory()).saveMessages({ messages });
  }
  getModelId() {
    return this.#modelId;
  }
  setModelId(modelId) {
    const previousModelId = this.#modelId;
    this.#modelId = modelId;
    if (modelId !== previousModelId) {
      this.#events.emit({ type: "model_changed", modelId, previousModelId });
    }
  }
  getMode() {
    return this.#mode;
  }
  setMode(mode) {
    const previousModeId = this.#mode.id;
    this.#mode = mode;
    if (mode.id !== previousModeId) {
      this.#events.emit({ type: "mode_changed", modeId: mode.id, previousModeId });
    }
  }
  async #buildRequestContext(requestContext) {
    requestContext ??= new RequestContext();
    const harnessContext = {
      threadId: this.#threadId,
      resourceId: this.#resourceId,
      modeId: this.#mode.id
      // harnessId: this.id,
      // state: this.getState(),
      // getState: () => this.getState(),
      // setState: updates => this.setState(updates),
      // updateState: updater => this.updateState(updater),
      // threadId: this.currentThreadId,
      // resourceId: this.resourceId,
      // modeId: this.currentModeId,
      // abortSignal: this.abortController?.signal,
      // workspace: this.workspace,
      // emitEvent: event => this.emit(event),
      // registerQuestion: params => this.registerQuestion(params),
      // registerPlanApproval: params => this.registerPlanApproval(params),
      // getSubagentModelId: params => this.getSubagentModelId(params),
    };
    requestContext.set("harness", harnessContext);
    return requestContext;
  }
  async #resolveMemory() {
    const mem = this.#memory;
    if (!mem) {
      throw new Error("Memory is not configured on this Harness");
    }
    if (typeof mem !== "function") {
      return mem;
    }
    const requestContext = await this.#buildRequestContext();
    const resolved = await mem({ requestContext });
    if (!resolved) {
      throw new Error("Dynamic memory factory returned empty value");
    }
    return resolved;
  }
};

// src/harness/v1/harness.ts
var Harness = class {
  #ownerId;
  #defaultMode;
  #modesById = /* @__PURE__ */ new Map();
  #storage;
  #compositeStorage;
  #memory;
  #events;
  constructor(config) {
    if (!config.modes.length) {
      throw new Error("The harness needs modes to operate.");
    }
    this.#ownerId = config.ownerId ?? randomUUID();
    this.#defaultMode = config.defaultModeId ?? config.modes[0].id;
    this.#storage = config.storage;
    this.#compositeStorage = config.mastra?.getStorage();
    this.#memory = config.memory;
    this.#events = new EventEmitter();
    const modes = config.modes ?? [];
    for (const mode of modes) {
      if (this.#modesById.has(mode.id)) {
        throw new Error(`Duplicate mode id "${mode.id}" found when creating the Harness`);
      }
      if (mode.tools && mode.additionalTools) {
        throw new Error(`Mode "${mode.id} cannot set both "tools" and "additionalTools" - choose replace OR augment`);
      }
      this.#modesById.set(mode.id, mode);
    }
  }
  get ownerId() {
    return this.#ownerId;
  }
  subscribe(listener) {
    return this.#events.subscribe(listener);
  }
  emit(event) {
    return this.#events.emit(event);
  }
  listModes() {
    return [...this.#modesById.values()];
  }
  /**
   * Look up a single mode by id. Returns `undefined` if no mode with that id
   * is registered. For the throwing variant used during request resolution,
   * see the internal `_getMode` helper.
   */
  getMode(modeId) {
    return this.#modesById.get(modeId);
  }
  async listSessions() {
    const storage = await this.#requireStorage();
    return storage.listSessions();
  }
  async session(opts) {
    const storage = await this.#requireStorage();
    if ("threadId" in opts) {
      return this.#sessionByThread(storage, opts);
    }
    const record = await this.#loadSessionRecord(storage, opts.sessionId, opts.resourceId);
    return this.#sessionFromRecord(record);
  }
  async cloneSession(session, opts = {}) {
    const storage = await this.#requireStorage();
    const source = await this.#loadSessionRecord(storage, session.id, session.resourceId);
    const modeId = opts.modeId ?? source.modeId;
    const mode = this.#modesById.get(modeId);
    if (!mode) {
      throw new Error(`Harness session "${source.id}" cannot clone into unknown mode "${modeId}"`);
    }
    const clone = await session.clone({
      ...opts,
      resourceId: opts.resourceId ?? source.resourceId,
      mode,
      modelId: opts.modelId ?? source.modelId
    });
    const record = {
      ...source,
      id: clone.id,
      ownerId: this.#ownerId,
      threadId: clone.threadId,
      resourceId: clone.resourceId,
      parentSessionId: opts.parentSessionId ?? source.id,
      origin: opts.origin ?? source.origin,
      modeId: opts.modeId ?? source.modeId,
      modelId: opts.modelId ?? source.modelId
    };
    await storage.saveSession(record);
    this.#events.emit({ type: "session_created", ...sessionCreatedPayload(record) });
    return this.#sessionFromRecord(record);
  }
  async #sessionByThread(storage, opts) {
    const id = this.#sessionIdFor(opts.resourceId, opts.threadId);
    const existing = await storage.loadSession(id);
    if (existing) {
      return this.#sessionFromRecord(existing);
    }
    const modeId = opts.modeId ?? this.#defaultMode;
    const mode = this.#modesById.get(modeId);
    if (!mode) {
      throw new Error(`Harness session for thread "${opts.threadId}" cannot use unknown mode "${modeId}"`);
    }
    const record = {
      id,
      ownerId: this.#ownerId,
      threadId: opts.threadId,
      resourceId: opts.resourceId,
      origin: "top-level",
      modeId,
      modelId: opts.modelId ?? mode.defaultModelId,
      createdAt: /* @__PURE__ */ new Date(),
      lastActivityAt: /* @__PURE__ */ new Date()
    };
    await storage.saveSession(record);
    this.#events.emit({ type: "session_created", ...sessionCreatedPayload(record) });
    return this.#sessionFromRecord(record);
  }
  async #loadSessionRecord(storage, sessionId, resourceId) {
    const record = await storage.loadSession(sessionId);
    if (!record) {
      throw new Error(`Harness session "${sessionId}" was not found`);
    }
    if (resourceId && record.resourceId !== resourceId) {
      throw new Error(`Harness session "${sessionId}" does not belong to resource "${resourceId}"`);
    }
    return record;
  }
  async #requireStorage() {
    if (this.#storage) {
      return this.#storage;
    }
    const storage = await this.#compositeStorage?.getStore("harness");
    if (!storage) {
      throw new Error("Harness session storage is not configured");
    }
    return storage;
  }
  #sessionFromRecord(record) {
    const mode = record.modeId ? this.#modesById.get(record.modeId) : this.#modesById.values().next().value;
    if (!mode) {
      throw new Error(`Harness session "${record.id}" references unknown mode "${record.modeId}"`);
    }
    return new Session({
      id: record.id,
      ownerId: record.ownerId,
      threadId: record.threadId,
      resourceId: record.resourceId,
      mode,
      model: record.modelId,
      createdAt: record.createdAt,
      lastActivityAt: record.lastActivityAt,
      memory: this.#memory,
      events: this.#events.scoped({ sessionId: record.id })
    });
  }
  #sessionIdFor(resourceId, threadId) {
    const hash = createHash("sha256").update(`${resourceId}\0${threadId}`).digest("hex").slice(0, 32);
    return `sess-${hash}`;
  }
};

export { EventEmitter, Harness, Session, formatHarnessEventId, parseHarnessEventId };
