// @mastra/core/dist/storage/constants.cjs
'use strict';

var chunkWZHQJ2FB_cjs = require('../chunk-WZHQJ2FB.cjs');



Object.defineProperty(exports, "AGENTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.AGENTS_SCHEMA; }
});
Object.defineProperty(exports, "AGENT_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.AGENT_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "DATASETS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.DATASETS_SCHEMA; }
});
Object.defineProperty(exports, "DATASET_ITEMS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.DATASET_ITEMS_SCHEMA; }
});
Object.defineProperty(exports, "DATASET_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.DATASET_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "EXPERIMENTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.EXPERIMENTS_SCHEMA; }
});
Object.defineProperty(exports, "EXPERIMENT_RESULTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.EXPERIMENT_RESULTS_SCHEMA; }
});
Object.defineProperty(exports, "FAVORITES_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.FAVORITES_SCHEMA; }
});
Object.defineProperty(exports, "MCP_CLIENTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.MCP_CLIENTS_SCHEMA; }
});
Object.defineProperty(exports, "MCP_CLIENT_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.MCP_CLIENT_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "MCP_SERVERS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.MCP_SERVERS_SCHEMA; }
});
Object.defineProperty(exports, "MCP_SERVER_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.MCP_SERVER_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.OBSERVATIONAL_MEMORY_SCHEMA; }
});
Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_TABLE_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.OBSERVATIONAL_MEMORY_TABLE_SCHEMA; }
});
Object.defineProperty(exports, "OLD_SPAN_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.OLD_SPAN_SCHEMA; }
});
Object.defineProperty(exports, "PROMPT_BLOCKS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.PROMPT_BLOCKS_SCHEMA; }
});
Object.defineProperty(exports, "PROMPT_BLOCK_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.PROMPT_BLOCK_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "SCORERS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SCORERS_SCHEMA; }
});
Object.defineProperty(exports, "SCORER_DEFINITIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SCORER_DEFINITIONS_SCHEMA; }
});
Object.defineProperty(exports, "SCORER_DEFINITION_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SCORER_DEFINITION_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "SKILLS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SKILLS_SCHEMA; }
});
Object.defineProperty(exports, "SKILL_BLOBS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SKILL_BLOBS_SCHEMA; }
});
Object.defineProperty(exports, "SKILL_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SKILL_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "SPAN_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.SPAN_SCHEMA; }
});
Object.defineProperty(exports, "TABLE_AGENTS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_AGENTS; }
});
Object.defineProperty(exports, "TABLE_AGENT_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_AGENT_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_BACKGROUND_TASKS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_BACKGROUND_TASKS; }
});
Object.defineProperty(exports, "TABLE_CHANNEL_CONFIG", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_CHANNEL_CONFIG; }
});
Object.defineProperty(exports, "TABLE_CHANNEL_INSTALLATIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_CHANNEL_INSTALLATIONS; }
});
Object.defineProperty(exports, "TABLE_CONFIGS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_CONFIGS; }
});
Object.defineProperty(exports, "TABLE_DATASETS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_DATASETS; }
});
Object.defineProperty(exports, "TABLE_DATASET_ITEMS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_DATASET_ITEMS; }
});
Object.defineProperty(exports, "TABLE_DATASET_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_DATASET_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_EXPERIMENTS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_EXPERIMENTS; }
});
Object.defineProperty(exports, "TABLE_EXPERIMENT_RESULTS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_EXPERIMENT_RESULTS; }
});
Object.defineProperty(exports, "TABLE_FAVORITES", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_FAVORITES; }
});
Object.defineProperty(exports, "TABLE_MCP_CLIENTS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_MCP_CLIENTS; }
});
Object.defineProperty(exports, "TABLE_MCP_CLIENT_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_MCP_CLIENT_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_MCP_SERVERS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_MCP_SERVERS; }
});
Object.defineProperty(exports, "TABLE_MCP_SERVER_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_MCP_SERVER_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_MESSAGES", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_MESSAGES; }
});
Object.defineProperty(exports, "TABLE_OBSERVATIONAL_MEMORY", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_OBSERVATIONAL_MEMORY; }
});
Object.defineProperty(exports, "TABLE_PROMPT_BLOCKS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_PROMPT_BLOCKS; }
});
Object.defineProperty(exports, "TABLE_PROMPT_BLOCK_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_PROMPT_BLOCK_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_RESOURCES", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_RESOURCES; }
});
Object.defineProperty(exports, "TABLE_SCHEDULES", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SCHEDULES; }
});
Object.defineProperty(exports, "TABLE_SCHEDULE_TRIGGERS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SCHEDULE_TRIGGERS; }
});
Object.defineProperty(exports, "TABLE_SCHEMAS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SCHEMAS; }
});
Object.defineProperty(exports, "TABLE_SCORERS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SCORERS; }
});
Object.defineProperty(exports, "TABLE_SCORER_DEFINITIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SCORER_DEFINITIONS; }
});
Object.defineProperty(exports, "TABLE_SCORER_DEFINITION_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SCORER_DEFINITION_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_SKILLS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SKILLS; }
});
Object.defineProperty(exports, "TABLE_SKILL_BLOBS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SKILL_BLOBS; }
});
Object.defineProperty(exports, "TABLE_SKILL_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SKILL_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_SPANS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_SPANS; }
});
Object.defineProperty(exports, "TABLE_THREADS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_THREADS; }
});
Object.defineProperty(exports, "TABLE_TOOL_PROVIDER_CONNECTIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_TOOL_PROVIDER_CONNECTIONS; }
});
Object.defineProperty(exports, "TABLE_TRACES", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_TRACES; }
});
Object.defineProperty(exports, "TABLE_WORKFLOW_SNAPSHOT", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_WORKFLOW_SNAPSHOT; }
});
Object.defineProperty(exports, "TABLE_WORKSPACES", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_WORKSPACES; }
});
Object.defineProperty(exports, "TABLE_WORKSPACE_VERSIONS", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TABLE_WORKSPACE_VERSIONS; }
});
Object.defineProperty(exports, "TOOL_PROVIDER_CONNECTIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.TOOL_PROVIDER_CONNECTIONS_SCHEMA; }
});
Object.defineProperty(exports, "WORKSPACES_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.WORKSPACES_SCHEMA; }
});
Object.defineProperty(exports, "WORKSPACE_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkWZHQJ2FB_cjs.WORKSPACE_VERSIONS_SCHEMA; }
});
