// @mastra/core/dist/harness/index.cjs
'use strict';

var chunk2IPLS64X_cjs = require('../chunk-2IPLS64X.cjs');
var chunkPPP5XYB5_cjs = require('../chunk-PPP5XYB5.cjs');
var chunk5SOAYQMP_cjs = require('../chunk-5SOAYQMP.cjs');
var chunkNLACGSDP_cjs = require('../chunk-NLACGSDP.cjs');
var chunk7Y2NL3JH_cjs = require('../chunk-7Y2NL3JH.cjs');
var chunkXB4FLS7A_cjs = require('../chunk-XB4FLS7A.cjs');
var chunkDCCAHUCG_cjs = require('../chunk-DCCAHUCG.cjs');
var chunkVPZGHUNF_cjs = require('../chunk-VPZGHUNF.cjs');
var chunk64VGLWFD_cjs = require('../chunk-64VGLWFD.cjs');
var crypto = require('crypto');
var v4 = require('zod/v4');

// src/harness/display-state-scheduler.ts
var DEFAULT_DISPLAY_STATE_SUBSCRIPTION_OPTIONS = {
  windowMs: 250,
  maxWaitMs: 500
};
var CRITICAL_DISPLAY_STATE_EVENT_TYPES = /* @__PURE__ */ new Set([
  "agent_start",
  "agent_end",
  "follow_up_queued",
  "error",
  "tool_approval_required",
  "tool_suspended",
  "ask_question",
  "plan_approval_required",
  "plan_approved",
  "thread_changed",
  "thread_created",
  "thread_deleted",
  "mode_changed",
  "model_changed",
  "subagent_model_changed",
  "state_changed",
  "tool_input_end",
  "tool_end",
  "subagent_end"
]);
function cloneValue(value, seen = /* @__PURE__ */ new WeakMap()) {
  if (value === null || typeof value !== "object") {
    return value;
  }
  if (value instanceof Date) {
    return new Date(value.getTime());
  }
  if (seen.has(value)) {
    return seen.get(value);
  }
  if (Array.isArray(value)) {
    const cloned2 = [];
    seen.set(value, cloned2);
    for (const item of value) {
      cloned2.push(cloneValue(item, seen));
    }
    return cloned2;
  }
  if (value instanceof Map) {
    const cloned2 = /* @__PURE__ */ new Map();
    seen.set(value, cloned2);
    for (const [key, mapValue] of value) {
      cloned2.set(cloneValue(key, seen), cloneValue(mapValue, seen));
    }
    return cloned2;
  }
  if (value instanceof Set) {
    const cloned2 = /* @__PURE__ */ new Set();
    seen.set(value, cloned2);
    for (const item of value) {
      cloned2.add(cloneValue(item, seen));
    }
    return cloned2;
  }
  const cloned = {};
  seen.set(value, cloned);
  for (const key of Reflect.ownKeys(value)) {
    cloned[key] = cloneValue(value[key], seen);
  }
  return cloned;
}
function cloneUnknown(value) {
  return cloneValue(value);
}
function cloneDisplayState(state) {
  return {
    ...state,
    currentMessage: state.currentMessage ? {
      ...state.currentMessage,
      createdAt: new Date(state.currentMessage.createdAt.getTime()),
      content: state.currentMessage.content.map((part) => cloneUnknown(part))
    } : null,
    tokenUsage: { ...state.tokenUsage },
    activeTools: new Map(
      Array.from(state.activeTools, ([id, tool]) => [
        id,
        {
          ...tool,
          args: cloneUnknown(tool.args),
          result: cloneUnknown(tool.result)
        }
      ])
    ),
    toolInputBuffers: new Map(Array.from(state.toolInputBuffers, ([id, buffer]) => [id, { ...buffer }])),
    pendingApproval: state.pendingApproval ? { ...state.pendingApproval, args: cloneUnknown(state.pendingApproval.args) } : null,
    pendingSuspension: state.pendingSuspension ? {
      ...state.pendingSuspension,
      args: cloneUnknown(state.pendingSuspension.args),
      suspendPayload: cloneUnknown(state.pendingSuspension.suspendPayload)
    } : null,
    pendingQuestion: state.pendingQuestion ? {
      ...state.pendingQuestion,
      options: state.pendingQuestion.options?.map((option) => cloneUnknown(option))
    } : null,
    pendingPlanApproval: state.pendingPlanApproval ? { ...state.pendingPlanApproval } : null,
    activeSubagents: new Map(
      Array.from(state.activeSubagents, ([id, subagent]) => [
        id,
        {
          ...subagent,
          toolCalls: subagent.toolCalls.map((toolCall) => cloneUnknown(toolCall))
        }
      ])
    ),
    omProgress: {
      ...state.omProgress,
      buffered: {
        observations: { ...state.omProgress.buffered.observations },
        reflection: { ...state.omProgress.buffered.reflection }
      }
    },
    modifiedFiles: new Map(
      Array.from(state.modifiedFiles, ([path, modifiedFile]) => [
        path,
        {
          ...modifiedFile,
          firstModified: new Date(modifiedFile.firstModified.getTime()),
          operations: [...modifiedFile.operations]
        }
      ])
    ),
    tasks: state.tasks.map((task) => cloneUnknown(task)),
    previousTasks: state.previousTasks.map((task) => cloneUnknown(task))
  };
}
var DisplayStateScheduler = class {
  constructor(listener, windowMs, maxWaitMs) {
    this.listener = listener;
    this.windowMs = windowMs;
    this.maxWaitMs = maxWaitMs;
  }
  listener;
  windowMs;
  maxWaitMs;
  disposed = false;
  pendingState = null;
  windowTimer = null;
  maxWaitTimer = null;
  notify(state, isCritical) {
    if (this.disposed) return;
    if (isCritical) {
      this.flush(state);
      return;
    }
    this.pendingState = state;
    if (this.windowTimer) {
      clearTimeout(this.windowTimer);
    }
    this.windowTimer = setTimeout(() => this.flushPending(), this.windowMs);
    if (!this.maxWaitTimer) {
      this.maxWaitTimer = setTimeout(() => this.flushPending(), this.maxWaitMs);
    }
  }
  dispose() {
    this.disposed = true;
    this.pendingState = null;
    this.clearTimers();
  }
  flushPending() {
    if (!this.pendingState) return;
    this.flush(this.pendingState);
  }
  flush(state) {
    if (this.disposed) return;
    this.pendingState = null;
    this.clearTimers();
    try {
      const result = this.listener(cloneDisplayState(state));
      if (result && typeof result === "object" && "catch" in result && typeof result.catch === "function") {
        result.catch((err) => console.error("Error in harness display state listener:", err));
      }
    } catch (err) {
      console.error("Error in harness display state listener:", err);
    }
  }
  clearTimers() {
    if (this.windowTimer) {
      clearTimeout(this.windowTimer);
    }
    if (this.maxWaitTimer) {
      clearTimeout(this.maxWaitTimer);
    }
    this.windowTimer = null;
    this.maxWaitTimer = null;
  }
};
var questionCounter = 0;
var planCounter = 0;
var FORKED_SUBAGENT_NESTING_NOTICE = "Do not call the `subagent` tool. You are currently running inside a forked subagent, and this is the maximum allowed subagent nesting level. Further subagent calls will return an error. Answer the task directly using the conversation history and the other tools available to you.";
var FORKED_SUBAGENT_TASK_NOTICE = "Do not call `task_write`, `task_update`, `task_complete`, or `task_check` inside a forked subagent. Forked subagents keep the parent task-tool schemas for prompt-cache stability, but the parent agent owns the visible task list. Track forked subagent work in your final answer instead.";
function formatQuestionAnswer(answer) {
  return Array.isArray(answer) ? answer.join(", ") : answer;
}
var askUserTool = chunk7Y2NL3JH_cjs.createTool({
  id: "ask_user",
  description: "Ask the user a question and wait for their response. Use this when you need clarification, want to validate assumptions, or need the user to make a decision between options. Provide options for structured choices (2-4 options), or omit them for open-ended questions. Use selectionMode to choose whether the user can pick one option or multiple options.",
  inputSchema: v4.z.object({
    question: v4.z.string().min(1).describe("The question to ask the user. Should be clear and specific."),
    options: v4.z.array(
      v4.z.object({
        label: v4.z.string().describe("Short display text for this option (1-5 words)"),
        description: v4.z.string().optional().describe("Explanation of what this option means")
      })
    ).optional().describe("Optional choices. If provided, shows a selection list. If omitted, shows a free-text input."),
    selectionMode: v4.z.enum(["single_select", "multi_select"]).optional().describe(
      "Controls how many provided options the user can select. Defaults to single_select when options are provided. Requires options."
    )
  }),
  execute: async ({ question, options, selectionMode }, context) => {
    try {
      const harnessCtx = context?.requestContext?.get("harness");
      const resolvedSelectionMode = options?.length ? selectionMode ?? "single_select" : void 0;
      if (selectionMode && !options?.length) {
        return {
          content: "Failed to ask user: selectionMode requires options.",
          isError: true
        };
      }
      if (!harnessCtx?.emitEvent || !harnessCtx?.registerQuestion) {
        return {
          content: `[Question for user]: ${question}${options?.length ? "\nOptions: " + options.map((o) => o.label).join(", ") : ""}${resolvedSelectionMode ? "\nSelection mode: " + resolvedSelectionMode : ""}`,
          isError: false
        };
      }
      const questionId = `q_${++questionCounter}_${Date.now()}`;
      const answer = await new Promise((resolve, reject) => {
        const signal = harnessCtx.abortSignal;
        if (signal?.aborted) {
          reject(new DOMException("Aborted", "AbortError"));
          return;
        }
        const onAbort = () => reject(new DOMException("Aborted", "AbortError"));
        signal?.addEventListener("abort", onAbort, { once: true });
        harnessCtx.registerQuestion({
          questionId,
          resolve: (answer2) => {
            signal?.removeEventListener("abort", onAbort);
            resolve(answer2);
          }
        });
        harnessCtx.emitEvent({
          type: "ask_question",
          questionId,
          question,
          options,
          selectionMode: resolvedSelectionMode
        });
      });
      return { content: `User answered: ${formatQuestionAnswer(answer)}`, isError: false };
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return { content: `Failed to ask user: ${msg}`, isError: true };
    }
  }
});
var submitPlanTool = chunk7Y2NL3JH_cjs.createTool({
  id: "submit_plan",
  description: "Submit a completed implementation plan for user review. The plan will be rendered as markdown and the user can approve, reject, or request changes. Use this when your exploration is complete and you have a concrete plan ready for review. On approval, the system automatically switches to the default mode so you can implement.",
  inputSchema: v4.z.object({
    title: v4.z.string().optional().describe("Short title for the plan (e.g., 'Add dark mode toggle')"),
    plan: v4.z.string().min(1).describe("The full plan content in markdown format. Should include Overview, Steps, and Verification sections.")
  }),
  execute: async ({ title, plan }, context) => {
    try {
      const harnessCtx = context?.requestContext?.get("harness");
      if (!harnessCtx?.emitEvent || !harnessCtx?.registerPlanApproval) {
        return {
          content: `[Plan submitted for review]

Title: ${title || "Implementation Plan"}

${plan}`,
          isError: false
        };
      }
      const planId = `plan_${++planCounter}_${Date.now()}`;
      const result = await new Promise((resolve, reject) => {
        const signal = harnessCtx.abortSignal;
        if (signal?.aborted) {
          reject(new DOMException("Aborted", "AbortError"));
          return;
        }
        const onAbort = () => reject(new DOMException("Aborted", "AbortError"));
        signal?.addEventListener("abort", onAbort, { once: true });
        harnessCtx.registerPlanApproval({
          planId,
          resolve: (res) => {
            signal?.removeEventListener("abort", onAbort);
            resolve(res);
          }
        });
        harnessCtx.emitEvent({
          type: "plan_approval_required",
          planId,
          title: title || "Implementation Plan",
          plan
        });
      });
      if (result.action === "approved") {
        return {
          content: "Plan approved. Proceed with implementation following the approved plan.",
          isError: false
        };
      }
      const feedback = result.feedback ? `

User feedback: ${result.feedback}` : "";
      return {
        content: `Plan was not approved. The user wants revisions.${feedback}

Please revise the plan based on the feedback and submit again with submit_plan.`,
        isError: false
      };
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return { content: `Failed to submit plan: ${msg}`, isError: true };
    }
  }
});
var taskIdSchema = v4.z.string().min(1).describe("Stable task identifier (for example, 'task_investigate_tests'). Keep this unchanged across updates.");
var taskItemInputSchema = v4.z.object({
  id: taskIdSchema.optional(),
  content: v4.z.string().min(1).describe("Task description in imperative form (e.g., 'Fix authentication bug')"),
  status: v4.z.enum(["pending", "in_progress", "completed"]).describe("Current task status"),
  activeForm: v4.z.string().min(1).describe("Present continuous form shown during execution (e.g., 'Fixing authentication bug')")
});
var taskItemSchema = taskItemInputSchema.extend({
  id: taskIdSchema
});
var taskToolResultSchema = v4.z.object({
  content: v4.z.string(),
  tasks: v4.z.array(taskItemSchema),
  isError: v4.z.boolean()
});
var taskCheckSummarySchema = v4.z.object({
  total: v4.z.number().int().nonnegative(),
  completed: v4.z.number().int().nonnegative(),
  inProgress: v4.z.number().int().nonnegative(),
  pending: v4.z.number().int().nonnegative(),
  incomplete: v4.z.number().int().nonnegative(),
  hasTasks: v4.z.boolean(),
  allCompleted: v4.z.boolean()
});
var taskCheckResultSchema = taskToolResultSchema.extend({
  summary: taskCheckSummarySchema,
  incompleteTasks: v4.z.array(taskItemSchema)
});
var TASK_ID_SLUG_MAX_LENGTH = 48;
function slugifyTaskContent(content) {
  let slug = "";
  let pendingSeparator = false;
  for (const char of content.toLowerCase()) {
    const code = char.charCodeAt(0);
    const isAsciiLetter = code >= 97 && code <= 122;
    const isDigit = code >= 48 && code <= 57;
    if (isAsciiLetter || isDigit) {
      if (pendingSeparator && slug.length > 0 && slug.length < TASK_ID_SLUG_MAX_LENGTH) {
        slug += "_";
      }
      if (slug.length >= TASK_ID_SLUG_MAX_LENGTH) break;
      slug += char;
      pendingSeparator = false;
      continue;
    }
    pendingSeparator = slug.length > 0;
  }
  return slug;
}
function createDeterministicTaskId(task, occurrence) {
  const slug = slugifyTaskContent(task.content);
  const suffix = occurrence > 1 ? `_${occurrence}` : "";
  return `task_${slug || "item"}${suffix}`;
}
function makeUniqueTaskId(id, usedIds, reservedIds = /* @__PURE__ */ new Set()) {
  if (!usedIds.has(id) && !reservedIds.has(id)) return id;
  let suffix = 2;
  let nextId = `${id}_${suffix}`;
  while (usedIds.has(nextId) || reservedIds.has(nextId)) {
    suffix += 1;
    nextId = `${id}_${suffix}`;
  }
  return nextId;
}
function assignTaskIds(tasks, previousTasks = []) {
  const usedIds = /* @__PURE__ */ new Set();
  const contentOccurrences = /* @__PURE__ */ new Map();
  const omittedContentCounts = /* @__PURE__ */ new Map();
  const explicitTaskIds = new Set(tasks.map((task) => task.id).filter((id) => Boolean(id)));
  const reusablePreviousIds = /* @__PURE__ */ new Map();
  for (const task of tasks) {
    if (!task.id) {
      omittedContentCounts.set(task.content, (omittedContentCounts.get(task.content) ?? 0) + 1);
    }
  }
  tasks.forEach((task, index) => {
    if (task.id || omittedContentCounts.get(task.content) !== 1) return;
    const previousMatches = previousTasks.filter(
      (previous) => previous.content === task.content && !explicitTaskIds.has(previous.id)
    );
    if (previousMatches.length === 1) {
      reusablePreviousIds.set(index, previousMatches[0].id);
    }
  });
  const reservedIds = /* @__PURE__ */ new Set([...explicitTaskIds, ...reusablePreviousIds.values()]);
  return tasks.map((task, index) => {
    const contentOccurrence = (contentOccurrences.get(task.content) ?? 0) + 1;
    contentOccurrences.set(task.content, contentOccurrence);
    const fallbackId = createDeterministicTaskId(task, contentOccurrence);
    const reusablePreviousId = reusablePreviousIds.get(index);
    const requestedId = task.id && !usedIds.has(task.id) ? task.id : void 0;
    const id = requestedId ?? (reusablePreviousId && !usedIds.has(reusablePreviousId) ? reusablePreviousId : makeUniqueTaskId(fallbackId, usedIds, reservedIds));
    usedIds.add(id);
    return {
      id,
      content: task.content,
      status: task.status,
      activeForm: task.activeForm
    };
  });
}
function getTasksFromState(state) {
  const typedState = state;
  return assignTaskIds(typedState?.tasks || []);
}
function getCurrentTasks(harnessCtx) {
  const state = harnessCtx?.getState ? harnessCtx.getState() : harnessCtx?.state;
  return getTasksFromState(state);
}
async function readTasks(harnessCtx) {
  if (harnessCtx.updateState) {
    return harnessCtx.updateState((state) => ({
      result: getTasksFromState(state)
    }));
  }
  return getCurrentTasks(harnessCtx);
}
function formatTaskListResult(tasks) {
  const completed = tasks.filter((t) => t.status === "completed").length;
  const inProgress = tasks.find((t) => t.status === "in_progress");
  const total = tasks.length;
  let summary = `Tasks updated: [${completed}/${total} completed]`;
  if (inProgress) {
    summary += `
Currently: ${inProgress.activeForm} (${inProgress.id})`;
  }
  if (tasks.length > 0) {
    summary += `
Task IDs:
${tasks.map((t) => `- ${t.id}: ${t.content} (${t.status})`).join("\n")}`;
  }
  return summary;
}
function summarizeTaskCheck(tasks) {
  const completedTasks = tasks.filter((task) => task.status === "completed");
  const inProgressTasks = tasks.filter((task) => task.status === "in_progress");
  const pendingTasks = tasks.filter((task) => task.status === "pending");
  const incompleteTasks = [...inProgressTasks, ...pendingTasks];
  return {
    summary: {
      total: tasks.length,
      completed: completedTasks.length,
      inProgress: inProgressTasks.length,
      pending: pendingTasks.length,
      incomplete: incompleteTasks.length,
      hasTasks: tasks.length > 0,
      allCompleted: tasks.length > 0 && incompleteTasks.length === 0
    },
    inProgressTasks,
    pendingTasks,
    incompleteTasks
  };
}
function formatTaskCheckResult(taskCheck) {
  const { summary, inProgressTasks, pendingTasks } = taskCheck;
  if (!summary.hasTasks) {
    return "No tasks found. Consider using task_write to create a task list for complex work.";
  }
  let response = `Task Status: [${summary.completed}/${summary.total} completed]
`;
  response += `- Completed: ${summary.completed}
`;
  response += `- In Progress: ${summary.inProgress}
`;
  response += `- Pending: ${summary.pending}
`;
  response += `
All tasks completed: ${summary.allCompleted ? "YES" : "NO"}`;
  if (!summary.allCompleted) {
    response += "\n\nIncomplete tasks:";
    if (inProgressTasks.length > 0) {
      response += "\n\nIn Progress:";
      inProgressTasks.forEach((t) => {
        response += `
- ${t.id}: ${t.content}`;
      });
    }
    if (pendingTasks.length > 0) {
      response += "\n\nPending:";
      pendingTasks.forEach((t) => {
        response += `
- ${t.id}: ${t.content}`;
      });
    }
    response += "\n\nContinue working on these tasks before ending.";
  }
  return response;
}
function hasMultipleInProgress(tasks) {
  return tasks.filter((task) => task.status === "in_progress").length > 1;
}
function multipleInProgressError(tasks) {
  return {
    content: "Only one task can be in_progress at a time.",
    tasks,
    isError: true
  };
}
function demoteExtraInProgress(tasks, preferredIndex) {
  const inProgressIndices = tasks.reduce((acc, t, i) => {
    if (t.status === "in_progress") acc.push(i);
    return acc;
  }, []);
  if (inProgressIndices.length <= 1) return tasks;
  const keepIndex = preferredIndex !== void 0 && inProgressIndices.includes(preferredIndex) ? preferredIndex : inProgressIndices[inProgressIndices.length - 1];
  return tasks.map(
    (t, i) => t.status === "in_progress" && i !== keepIndex ? { ...t, status: "pending" } : t
  );
}
async function writeTasks(harnessCtx, tasks) {
  if (!harnessCtx) return;
  await harnessCtx.setState({ tasks });
  harnessCtx.emitEvent?.({
    type: "task_updated",
    tasks
  });
}
async function mutateTasks(harnessCtx, mutation) {
  if (harnessCtx.updateState) {
    return harnessCtx.updateState((state) => {
      const result2 = mutation(getTasksFromState(state));
      return {
        result: result2,
        updates: result2.isError ? void 0 : { tasks: result2.tasks },
        events: result2.isError ? void 0 : [
          {
            type: "task_updated",
            tasks: result2.tasks
          }
        ]
      };
    });
  }
  const result = mutation(getCurrentTasks(harnessCtx));
  if (!result.isError) {
    await writeTasks(harnessCtx, result.tasks);
  }
  return result;
}
function formatAvailableTaskIds(tasks) {
  if (tasks.length === 0) return "No tasks are currently tracked.";
  return `Available task IDs:
${tasks.map((t) => `- ${t.id}: ${t.content} (${t.status})`).join("\n")}`;
}
var taskWriteTool = chunk7Y2NL3JH_cjs.createTool({
  id: "task_write",
  description: `Create and manage a structured task list for your current coding session. This helps you track progress, organize complex tasks, and demonstrate thoroughness to the user.

Usage:
- Use this to create the initial task list or replace the whole list after replanning
- Pass the FULL task list each time this tool is called (replaces the previous list)
- Each task has: id (stable identifier), content (imperative), status (pending, in_progress, or completed), activeForm (present continuous)
- IDs must be unique. If duplicate explicit IDs are provided, the duplicate task is returned with a generated fallback ID
- Keep task IDs stable across updates. If omitted, IDs are generated and returned in the tool result
- When an ID is omitted while rewriting an existing list, one unambiguous matching task may reuse an existing ID
- Prefer single-task update tools when they are available
- Mark tasks in_progress BEFORE starting work (only ONE at a time)
- Mark tasks completed IMMEDIATELY after finishing
- Use this for multi-step tasks requiring 3+ distinct actions

States:
- pending: Not yet started
- in_progress: Currently working on (limit to ONE)
- completed: Finished successfully`,
  inputSchema: v4.z.object({
    tasks: v4.z.array(taskItemInputSchema).describe("The complete updated task list")
  }),
  outputSchema: taskToolResultSchema,
  execute: async ({ tasks }, context) => {
    try {
      const harnessCtx = context?.requestContext?.get("harness");
      if (!harnessCtx) {
        return {
          content: "Unable to update task list (no harness context)",
          tasks: [],
          isError: true
        };
      }
      return await mutateTasks(harnessCtx, (currentTasks) => {
        const normalizedTasks = assignTaskIds(tasks, currentTasks);
        if (hasMultipleInProgress(normalizedTasks)) {
          return multipleInProgressError(currentTasks);
        }
        return {
          content: formatTaskListResult(normalizedTasks),
          tasks: normalizedTasks,
          isError: false
        };
      });
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return {
        content: `Failed to update tasks: ${msg}`,
        tasks: [],
        isError: true
      };
    }
  }
});
var taskUpdateTool = chunk7Y2NL3JH_cjs.createTool({
  id: "task_update",
  description: `Update one task in the current task list by stable ID. Use this for targeted changes to one existing task.

Usage:
- Provide the task ID returned by the task-list tools
- Include only the fields that changed
- Use status to move a task between pending, in_progress, and completed
- Use task_complete when only marking a task completed
- If the ID is unknown, the tool returns an error with available task IDs`,
  inputSchema: v4.z.object({
    id: taskIdSchema,
    content: v4.z.string().min(1).optional().describe("New task description in imperative form"),
    status: v4.z.enum(["pending", "in_progress", "completed"]).optional().describe("New task status"),
    activeForm: v4.z.string().min(1).optional().describe("New present continuous form shown during execution")
  }).refine((input) => input.content !== void 0 || input.status !== void 0 || input.activeForm !== void 0, {
    message: "Provide at least one field to update."
  }),
  outputSchema: taskToolResultSchema,
  execute: async ({ id, content, status, activeForm }, context) => {
    try {
      const harnessCtx = context?.requestContext?.get("harness");
      if (!harnessCtx) {
        return {
          content: "Unable to update task list (no harness context)",
          tasks: [],
          isError: true
        };
      }
      return await mutateTasks(harnessCtx, (tasks) => {
        const taskIndex = tasks.findIndex((task) => task.id === id);
        if (taskIndex === -1) {
          return {
            content: `Task not found: ${id}

${formatAvailableTaskIds(tasks)}`,
            tasks,
            isError: true
          };
        }
        const updatedTasks = demoteExtraInProgress(
          tasks.map(
            (task, index) => index === taskIndex ? {
              ...task,
              ...content !== void 0 ? { content } : {},
              ...status !== void 0 ? { status } : {},
              ...activeForm !== void 0 ? { activeForm } : {}
            } : task
          ),
          taskIndex
        );
        return {
          content: formatTaskListResult(updatedTasks),
          tasks: updatedTasks,
          isError: false
        };
      });
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return {
        content: `Failed to update task: ${msg}`,
        tasks: [],
        isError: true
      };
    }
  }
});
var taskCompleteTool = chunk7Y2NL3JH_cjs.createTool({
  id: "task_complete",
  description: `Mark one task completed by stable ID. Use this when one tracked task is finished.

Usage:
- Provide the task ID returned by the task-list tools
- If the ID is unknown, the tool returns an error with available task IDs`,
  inputSchema: v4.z.object({
    id: taskIdSchema
  }),
  outputSchema: taskToolResultSchema,
  execute: async ({ id }, context) => {
    try {
      const harnessCtx = context?.requestContext?.get("harness");
      if (!harnessCtx) {
        return {
          content: "Unable to update task list (no harness context)",
          tasks: [],
          isError: true
        };
      }
      return await mutateTasks(harnessCtx, (tasks) => {
        const taskIndex = tasks.findIndex((task) => task.id === id);
        if (taskIndex === -1) {
          return {
            content: `Task not found: ${id}

${formatAvailableTaskIds(tasks)}`,
            tasks,
            isError: true
          };
        }
        const updatedTasks = tasks.map(
          (task, index) => index === taskIndex ? {
            ...task,
            status: "completed"
          } : task
        );
        return {
          content: formatTaskListResult(updatedTasks),
          tasks: updatedTasks,
          isError: false
        };
      });
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      return {
        content: `Failed to complete task: ${msg}`,
        tasks: [],
        isError: true
      };
    }
  }
});
var taskCheckTool = chunk7Y2NL3JH_cjs.createTool({
  id: "task_check",
  description: `Check the completion status of your current task list. Use this before finishing tracked work to ensure all tasks are completed.

Returns:
- Human-readable content summary with task counts and incomplete task IDs
- Structured task list snapshot with stable IDs
- summary object with total, completed, inProgress, pending, incomplete, hasTasks, and allCompleted
- incompleteTasks array for tasks that still need work

summary.allCompleted is true only when at least one tracked task exists and every tracked task is completed. If no tasks exist, summary.hasTasks is false and summary.allCompleted is false.`,
  inputSchema: v4.z.object({}),
  // No input needed
  outputSchema: taskCheckResultSchema,
  execute: async ({}, context) => {
    try {
      const harnessCtx = context?.requestContext?.get("harness");
      if (!harnessCtx) {
        const emptyCheck = summarizeTaskCheck([]);
        return {
          content: "Unable to access task list (no harness context)",
          tasks: [],
          summary: emptyCheck.summary,
          incompleteTasks: emptyCheck.incompleteTasks,
          isError: true
        };
      }
      const tasks = await readTasks(harnessCtx);
      const taskCheck = summarizeTaskCheck(tasks);
      return {
        content: formatTaskCheckResult(taskCheck),
        tasks,
        summary: taskCheck.summary,
        incompleteTasks: taskCheck.incompleteTasks,
        isError: false
      };
    } catch (error) {
      const msg = error instanceof Error ? error.message : "Unknown error";
      const emptyCheck = summarizeTaskCheck([]);
      return {
        content: `Failed to check tasks: ${msg}`,
        tasks: [],
        summary: emptyCheck.summary,
        incompleteTasks: emptyCheck.incompleteTasks,
        isError: true
      };
    }
  }
});
function createSubagentTool(opts) {
  const { subagents, resolveModel, harnessTools, fallbackModelId } = opts;
  const subagentIds = subagents.map((s) => s.id);
  const typeDescriptions = subagents.map((s) => `- **${s.id}** (${s.name}): ${s.description}`).join("\n");
  return chunk7Y2NL3JH_cjs.createTool({
    id: "subagent",
    description: `Delegate a focused task to a specialized subagent. The subagent runs independently with a constrained toolset, then returns its findings as text.

Available agent types:
${typeDescriptions}

By default the subagent runs in its own context \u2014 it does NOT see the parent conversation history. Write a clear, self-contained task description.

Set \`forked: true\` for context-dependent parallel work that needs the parent conversation, prior tool results, or the parent tool environment. Omit it for self-contained delegation. A forked subagent reuses the parent agent's instructions and tools so the prompt prefix stays cache-friendly.

Use this tool when:
- You want to run multiple investigations in parallel
- The task is self-contained and can be delegated`,
    inputSchema: v4.z.object({
      agentType: v4.z.enum(subagentIds).describe("Type of subagent to spawn"),
      task: v4.z.string().describe(
        "Clear, self-contained description of what the subagent should do. For non-forked subagents include all relevant context \u2014 the subagent cannot see the parent conversation."
      ),
      modelId: v4.z.string().optional().describe(
        "Optional model ID override for this task. Ignored when `forked: true` (the parent agent's model is used)."
      ),
      forked: v4.z.boolean().optional().describe(
        "If true, fork the parent conversation: clone the parent thread and run with the parent agent's instructions/tools so prompt cache is preserved. Requires memory to be configured on the Harness. Defaults to the subagent definition's `forked` setting."
      )
    }),
    execute: async (input, context) => {
      const { agentType, modelId, forked } = input;
      let { task } = input;
      const displayTask = task;
      const definition = subagents.find((s) => s.id === agentType);
      if (!definition) {
        return {
          content: `Unknown agent type: ${agentType}. Valid types: ${subagentIds.join(", ")}`,
          isError: true
        };
      }
      const harnessCtx = context?.requestContext?.get("harness");
      const emitEvent = harnessCtx?.emitEvent;
      const abortSignal = harnessCtx?.abortSignal;
      const toolCallId = context?.agent?.toolCallId ?? "unknown";
      const workspace = context?.workspace;
      const runAsForked = forked ?? definition.forked ?? false;
      let subagentToRun;
      let resolvedModelId;
      let subagentRequestContext;
      let streamMemory;
      let streamMaxSteps;
      let streamStopWhen;
      let streamPrepareStep;
      let forkedToolsets;
      if (runAsForked) {
        const parentAgent = opts.getParentAgent?.();
        if (!parentAgent) {
          return {
            content: "Forked subagent requires a parent agent. None is configured on this Harness.",
            isError: true
          };
        }
        const parentThreadId = harnessCtx?.threadId;
        if (!parentThreadId) {
          return {
            content: "Forked subagent requires an active parent thread; none is set on the Harness.",
            isError: true
          };
        }
        if (!opts.cloneThreadForFork) {
          return {
            content: "Forked subagent requires memory to be configured on the Harness so the parent thread can be cloned.",
            isError: true
          };
        }
        await context?.agent?.flushMessages?.().catch(() => {
        });
        let forkedThread;
        try {
          forkedThread = await opts.cloneThreadForFork({
            sourceThreadId: parentThreadId,
            resourceId: harnessCtx?.resourceId,
            title: `Fork: ${definition.name} subagent`
          });
        } catch (err) {
          return {
            content: `Failed to clone parent thread for forked subagent: ${err instanceof Error ? err.message : String(err)}`,
            isError: true
          };
        }
        subagentToRun = parentAgent;
        resolvedModelId = opts.getParentModelId?.() || "parent-agent";
        task = `${task}

${FORKED_SUBAGENT_NESTING_NOTICE}

${FORKED_SUBAGENT_TASK_NOTICE}`;
        streamMemory = { thread: forkedThread.id, resource: forkedThread.resourceId };
        streamMaxSteps = 1e3;
        streamStopWhen = void 0;
        streamPrepareStep = void 0;
        if (context?.requestContext) {
          subagentRequestContext = new chunk64VGLWFD_cjs.RequestContext(context.requestContext.entries());
          if (harnessCtx) {
            subagentRequestContext.set("harness", {
              ...harnessCtx,
              threadId: forkedThread.id,
              resourceId: forkedThread.resourceId
            });
          }
        }
        const inheritedToolsets = await opts.getParentToolsets?.(subagentRequestContext);
        if (inheritedToolsets) {
          forkedToolsets = {};
          for (const [setName, setTools] of Object.entries(inheritedToolsets)) {
            const patched = {};
            for (const [toolId, tool] of Object.entries(setTools)) {
              if (toolId === "subagent") {
                patched[toolId] = patchSubagentToolForFork(tool);
              } else if (isForkedTaskToolId(toolId)) {
                patched[toolId] = patchTaskToolForFork(tool, toolId);
              } else {
                patched[toolId] = tool;
              }
            }
            forkedToolsets[setName] = patched;
          }
        }
      } else {
        const mergedTools = { ...definition.tools };
        if (definition.allowedHarnessTools && harnessTools) {
          for (const toolId of definition.allowedHarnessTools) {
            if (harnessTools[toolId] && !mergedTools[toolId]) {
              mergedTools[toolId] = harnessTools[toolId];
            }
          }
        }
        const harnessModelId = harnessCtx?.getSubagentModelId?.({ agentType }) ?? void 0;
        const maybeModelId = modelId ?? harnessModelId ?? definition.defaultModelId ?? fallbackModelId;
        if (!maybeModelId) {
          return { content: "No model ID available for subagent. Configure defaultModelId.", isError: true };
        }
        resolvedModelId = maybeModelId;
        let model;
        try {
          model = resolveModel(resolvedModelId);
        } catch (err) {
          return {
            content: `Failed to resolve model "${resolvedModelId}": ${err instanceof Error ? err.message : String(err)}`,
            isError: true
          };
        }
        subagentToRun = new chunkPPP5XYB5_cjs.Agent({
          id: `subagent-${definition.id}`,
          name: `${definition.name} Subagent`,
          instructions: definition.instructions,
          model,
          tools: mergedTools,
          workspace
        });
        const allowedWs = definition.allowedWorkspaceTools ? new Set(definition.allowedWorkspaceTools) : void 0;
        const allWorkspaceToolNames = workspace && allowedWs ? new Set(
          Object.keys(
            await chunk5SOAYQMP_cjs.createWorkspaceTools(workspace, {
              requestContext: context?.requestContext ?? {},
              workspace
            })
          )
        ) : void 0;
        streamMaxSteps = definition.maxSteps ?? (definition.stopWhen ? void 0 : 50);
        streamStopWhen = definition.stopWhen;
        streamPrepareStep = allowedWs && allWorkspaceToolNames ? ({ tools }) => ({
          activeTools: Object.keys(tools ?? {}).filter((k) => !allWorkspaceToolNames.has(k) || allowedWs.has(k))
        }) : void 0;
        if (context?.requestContext) {
          subagentRequestContext = new chunk64VGLWFD_cjs.RequestContext(context.requestContext.entries());
          if (harnessCtx) {
            subagentRequestContext.set("harness", { ...harnessCtx, threadId: null, resourceId: "" });
          }
        }
      }
      const startTime = Date.now();
      emitEvent?.({
        type: "subagent_start",
        toolCallId,
        agentType,
        task: displayTask,
        modelId: resolvedModelId,
        forked: runAsForked
      });
      let partialText = "";
      try {
        const response = await subagentToRun.stream(task, {
          maxSteps: streamMaxSteps,
          stopWhen: streamStopWhen,
          abortSignal,
          requireToolApproval: false,
          requestContext: subagentRequestContext,
          ...streamMemory && { memory: streamMemory },
          ...forkedToolsets && { toolsets: forkedToolsets },
          ...context?.tracingContext && { tracingContext: context.tracingContext },
          prepareStep: streamPrepareStep
        });
        for await (const chunk of response.fullStream) {
          switch (chunk.type) {
            case "text-delta":
              partialText += chunk.payload.text;
              emitEvent?.({
                type: "subagent_text_delta",
                toolCallId,
                agentType,
                textDelta: chunk.payload.text
              });
              break;
            case "tool-call":
              if (!(runAsForked && chunk.payload.toolName === "subagent")) {
                emitEvent?.({
                  type: "subagent_tool_start",
                  toolCallId,
                  agentType,
                  subToolName: chunk.payload.toolName,
                  subToolArgs: chunk.payload.args
                });
              }
              break;
            case "tool-result": {
              const isErr = chunk.payload.isError ?? false;
              if (!(runAsForked && chunk.payload.toolName === "subagent")) {
                emitEvent?.({
                  type: "subagent_tool_end",
                  toolCallId,
                  agentType,
                  subToolName: chunk.payload.toolName,
                  subToolResult: chunk.payload.result,
                  isError: isErr
                });
              }
              break;
            }
          }
        }
        if (abortSignal?.aborted) {
          const durationMs2 = Date.now() - startTime;
          const abortResult = partialText ? `[Aborted by user]

Partial output:
${partialText}` : "[Aborted by user]";
          emitEvent?.({ type: "subagent_end", toolCallId, agentType, result: abortResult, isError: false, durationMs: durationMs2 });
          return { content: abortResult, isError: false };
        }
        const fullOutput = await response.getFullOutput();
        const resultText = fullOutput.text || partialText;
        const durationMs = Date.now() - startTime;
        emitEvent?.({ type: "subagent_end", toolCallId, agentType, result: resultText, isError: false, durationMs });
        return { content: resultText, isError: false };
      } catch (err) {
        const isAbort = err instanceof Error && (err.name === "AbortError" || err.message?.includes("abort") || err.message?.includes("cancel"));
        const durationMs = Date.now() - startTime;
        if (isAbort) {
          const abortResult = partialText ? `[Aborted by user]

Partial output:
${partialText}` : "[Aborted by user]";
          emitEvent?.({ type: "subagent_end", toolCallId, agentType, result: abortResult, isError: false, durationMs });
          return { content: abortResult, isError: false };
        }
        const message = err instanceof Error ? err.message : String(err);
        emitEvent?.({ type: "subagent_end", toolCallId, agentType, result: message, isError: true, durationMs });
        return { content: `Subagent "${definition.name}" failed: ${message}`, isError: true };
      }
    }
  });
}
function patchSubagentToolForFork(tool) {
  const stubExecute = async () => ({
    content: FORKED_SUBAGENT_NESTING_NOTICE,
    isError: true
  });
  return Object.assign({}, tool, { execute: stubExecute });
}
function isForkedTaskToolId(toolId) {
  return toolId === "task_write" || toolId === "task_update" || toolId === "task_complete" || toolId === "task_check";
}
function patchTaskToolForFork(tool, toolId) {
  const stubExecute = async () => {
    const baseResult = {
      content: FORKED_SUBAGENT_TASK_NOTICE,
      tasks: [],
      isError: true
    };
    if (toolId === "task_check") {
      const taskCheck = summarizeTaskCheck([]);
      return {
        ...baseResult,
        summary: taskCheck.summary,
        incompleteTasks: taskCheck.incompleteTasks
      };
    }
    return baseResult;
  };
  return Object.assign({}, tool, { execute: stubExecute });
}
function parseSubagentMeta(content) {
  const match = content.match(/\n<subagent-meta modelId="([^"]*)" durationMs="(\d+)" tools="([^"]*)" \/>$/);
  if (!match) return { text: content };
  const text = content.slice(0, match.index);
  const modelId = match[1];
  const durationMs = parseInt(match[2], 10);
  const toolCalls = match[3] ? match[3].split(",").filter(Boolean).map((entry) => {
    const [name, status] = entry.split(":");
    return { name, isError: status === "err" };
  }) : [];
  return { text, modelId, durationMs, toolCalls };
}

// src/harness/types.ts
function createEmptyTokenUsage() {
  return {
    promptTokens: 0,
    completionTokens: 0,
    totalTokens: 0,
    cachedInputTokens: 0,
    cacheCreationInputTokens: 0
  };
}
function defaultDisplayState() {
  return {
    isRunning: false,
    currentMessage: null,
    queuedFollowUps: 0,
    tokenUsage: createEmptyTokenUsage(),
    activeTools: /* @__PURE__ */ new Map(),
    toolInputBuffers: /* @__PURE__ */ new Map(),
    pendingApproval: null,
    pendingSuspension: null,
    pendingQuestion: null,
    pendingPlanApproval: null,
    activeSubagents: /* @__PURE__ */ new Map(),
    omProgress: defaultOMProgressState(),
    bufferingMessages: false,
    bufferingObservations: false,
    modifiedFiles: /* @__PURE__ */ new Map(),
    tasks: [],
    previousTasks: []
  };
}
function defaultOMProgressState() {
  return {
    status: "idle",
    pendingTokens: 0,
    threshold: 3e4,
    thresholdPercent: 0,
    observationTokens: 0,
    reflectionThreshold: 4e4,
    reflectionThresholdPercent: 0,
    buffered: {
      observations: {
        status: "idle",
        chunks: 0,
        messageTokens: 0,
        projectedMessageRemoval: 0,
        observationTokens: 0
      },
      reflection: {
        status: "idle",
        inputObservationTokens: 0,
        observationTokens: 0
      }
    },
    generationCount: 0,
    stepNumber: 0,
    preReflectionTokens: 0
  };
}

// src/harness/harness.ts
function getUsageNumber(usage, key) {
  const value = usage[key];
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "string" && value.trim() !== "") {
    const numericValue = Number(value);
    if (Number.isFinite(numericValue)) {
      return numericValue;
    }
  }
  return void 0;
}
function addOptionalUsageField(usage, key, value) {
  if (value !== void 0) {
    usage[key] = (usage[key] ?? 0) + value;
  }
}
function getDisplayTransform(metadata, phase, fallback) {
  const transform = chunkVPZGHUNF_cjs.getTransformedToolPayload(metadata, "display", phase);
  return chunkVPZGHUNF_cjs.hasTransformedToolPayload(transform) ? transform.transformed : fallback;
}
function getStringValue(value) {
  return typeof value === "string" ? value : void 0;
}
function getRecordValue(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : void 0;
}
function signalContentsToHarnessContent(contents) {
  if (typeof contents === "string") return [{ type: "text", text: contents }];
  return contents.flatMap((part) => {
    if (part.type === "text") {
      return [{ type: "text", text: part.text }];
    }
    if (typeof part.data !== "string") return [];
    if (part.mediaType.startsWith("image/")) {
      return [{ type: "image", data: part.data, mimeType: part.mediaType }];
    }
    return [
      {
        type: "file",
        data: part.data,
        mediaType: part.mediaType,
        filename: part.filename
      }
    ];
  });
}
function toSystemReminderContent(payload) {
  const attributes = getRecordValue(payload.attributes);
  const metadata = getRecordValue(payload.metadata);
  const message = getStringValue(payload.contents);
  if (message === void 0) return void 0;
  return {
    type: "system_reminder",
    message,
    reminderType: getStringValue(payload.reminderType) ?? getStringValue(attributes?.type) ?? getStringValue(payload.type),
    path: getStringValue(payload.path) ?? getStringValue(attributes?.path),
    precedesMessageId: getStringValue(payload.precedesMessageId) ?? getStringValue(attributes?.precedesMessageId),
    gapText: getStringValue(payload.gapText) ?? getStringValue(attributes?.gapText),
    gapMs: typeof payload.gapMs === "number" ? payload.gapMs : typeof attributes?.gapMs === "number" ? attributes.gapMs : void 0,
    timestamp: getStringValue(payload.timestamp) ?? getStringValue(attributes?.timestamp),
    goalMaxTurns: typeof payload.goalMaxTurns === "number" ? payload.goalMaxTurns : typeof metadata?.goalMaxTurns === "number" ? metadata.goalMaxTurns : void 0,
    judgeModelId: getStringValue(payload.judgeModelId) ?? getStringValue(metadata?.judgeModelId)
  };
}
function toUserSignalMessage(payload) {
  const id = getStringValue(payload.id);
  const rawContents = payload.contents;
  if (!id || rawContents === void 0) return void 0;
  const signal = chunkDCCAHUCG_cjs.createSignal({
    id,
    type: "user",
    tagName: "user",
    contents: rawContents,
    attributes: getRecordValue(payload.attributes),
    createdAt: getStringValue(payload.createdAt)
  });
  const content = signalContentsToHarnessContent(signal.contents);
  if (content.length === 0) return void 0;
  return {
    id: signal.id,
    role: "user",
    content,
    createdAt: signal.createdAt,
    attributes: signal.attributes
  };
}
var Harness = class {
  id;
  config;
  stateSchema;
  state;
  currentModeId;
  currentThreadId = null;
  resourceId;
  defaultResourceId;
  listeners = [];
  displayStateSchedulers = /* @__PURE__ */ new Set();
  abortController = null;
  abortRequested = false;
  currentRunId = null;
  currentTraceId = null;
  currentOperationId = 0;
  agentThreadSubscription = null;
  agentThreadSubscriptionKey = null;
  followUpQueue = [];
  pendingApprovalResolve = null;
  pendingApprovalToolName = null;
  pendingSuspensionRunId = null;
  pendingSuspensionToolCallId = null;
  pendingQuestions = /* @__PURE__ */ new Map();
  pendingPlanApprovals = /* @__PURE__ */ new Map();
  workspace = void 0;
  workspaceFn = void 0;
  workspaceInitialized = false;
  browser = void 0;
  browserFn = void 0;
  heartbeatTimers = /* @__PURE__ */ new Map();
  tokenUsage = createEmptyTokenUsage();
  sessionGrantedCategories = /* @__PURE__ */ new Set();
  sessionGrantedTools = /* @__PURE__ */ new Set();
  displayState = defaultDisplayState();
  stateUpdateQueue = Promise.resolve();
  switchModeVersion = 0;
  availableModelsCache = null;
  availableModelsCacheTime = 0;
  #internalMastra = void 0;
  constructor(config) {
    this.id = config.id;
    this.config = config;
    this.resourceId = config.resourceId ?? config.id;
    this.defaultResourceId = this.resourceId;
    this.stateSchema = config.stateSchema ? chunkXB4FLS7A_cjs.toStandardSchema(config.stateSchema) : void 0;
    this.state = {
      ...this.getSchemaDefaults(),
      ...config.initialState
    };
    const defaultMode = config.modes.find((m) => m.default) ?? config.modes[0];
    if (!defaultMode) {
      throw new Error("Harness requires at least one agent mode");
    }
    this.currentModeId = defaultMode.id;
    if (config.workspace instanceof chunk5SOAYQMP_cjs.Workspace) {
      this.workspace = config.workspace;
    } else if (typeof config.workspace === "function") {
      this.workspaceFn = config.workspace;
    }
    if (config.browser && typeof config.browser !== "function") {
      this.browser = config.browser;
    } else if (typeof config.browser === "function") {
      this.browserFn = config.browser;
    }
    const currentModel = this.state.currentModelId;
    if (!currentModel && defaultMode.defaultModelId) {
      void this.setState({ currentModelId: defaultMode.defaultModelId });
    }
  }
  // ===========================================================================
  // Accessors
  // ===========================================================================
  /**
   * Access the internal Mastra instance.
   * Available after `init()` when storage is configured.
   * Useful for scorer registration, observability access, and eval tooling.
   */
  getMastra() {
    return this.#internalMastra;
  }
  /**
   * Sets or updates the harness-level browser and propagates it to static mode agents.
   */
  setBrowser(browser) {
    this.browser = browser;
    this.browserFn = void 0;
    for (const mode of this.config.modes) {
      const agent = typeof mode.agent === "function" ? null : mode.agent;
      if (!agent || agent.hasOwnBrowser()) continue;
      agent.setBrowser(browser);
    }
  }
  // ===========================================================================
  // Initialization
  // ===========================================================================
  /**
   * Initialize the harness — loads storage and workspace.
   * Must be called before using the harness.
   */
  async init() {
    if (this.config.storage) {
      this.#internalMastra = new chunk2IPLS64X_cjs.Mastra({
        logger: false,
        storage: this.config.storage,
        ...this.config.pubsub ? { pubsub: this.config.pubsub } : {},
        ...this.config.observability ? { observability: this.config.observability } : {}
      });
      await this.#internalMastra.getStorage().init();
    }
    if (this.config.workspace && !this.workspaceInitialized && !this.workspaceFn) {
      try {
        if (!this.workspace) {
          this.workspace = new chunk5SOAYQMP_cjs.Workspace(this.config.workspace);
        }
        this.emit({ type: "workspace_status_changed", status: "initializing" });
        await this.workspace.init();
        this.workspaceInitialized = true;
        this.emit({ type: "workspace_status_changed", status: "ready" });
        this.emit({
          type: "workspace_ready",
          workspaceId: this.workspace.id,
          workspaceName: this.workspace.name
        });
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        this.workspace = void 0;
        this.workspaceInitialized = false;
        this.emit({ type: "workspace_status_changed", status: "error", error: err });
        this.emit({ type: "workspace_error", error: err });
      }
    }
    for (const mode of this.config.modes) {
      const agent = typeof mode.agent === "function" ? null : mode.agent;
      if (!agent) continue;
      this.propagateRuntimeServicesToAgent(agent);
    }
    this.startHeartbeats();
  }
  /**
   * Select the most recent thread, or create one if none exist.
   */
  async selectOrCreateThread() {
    const threads = await this.listThreads();
    if (threads.length === 0) {
      return await this.createThread();
    }
    const sortedThreads = [...threads].sort((a, b) => b.updatedAt.getTime() - a.updatedAt.getTime());
    const mostRecent = sortedThreads[0];
    await this.config.threadLock?.acquire(mostRecent.id);
    this.currentThreadId = mostRecent.id;
    await this.loadThreadMetadata();
    await this.ensureCurrentAgentThreadSubscription();
    return mostRecent;
  }
  async getMemoryStorage() {
    if (!this.config.storage) {
      throw new Error("Storage is not configured on this Harness");
    }
    const memoryStorage = await this.config.storage.getStore("memory");
    if (!memoryStorage) {
      throw new Error("Storage does not have a memory domain configured");
    }
    return memoryStorage;
  }
  // ===========================================================================
  // State Management
  // ===========================================================================
  /**
   * Get current harness state (read-only snapshot).
   */
  getState() {
    return { ...this.state };
  }
  async applyStateUpdates(updates) {
    const changedKeys = Object.keys(updates);
    const newState = { ...this.state, ...updates };
    if (this.stateSchema) {
      const result = await this.stateSchema["~standard"].validate(newState);
      if (result.issues) {
        const messages = result.issues.map((i) => i.message).join("; ");
        throw new Error(`Invalid state update: ${messages}`);
      }
      this.state = result.value;
    } else {
      this.state = newState;
    }
    this.emit({ type: "state_changed", state: this.state, changedKeys });
  }
  /**
   * Update harness state. Validates against schema if provided.
   * Emits state_changed event.
   */
  async setState(updates) {
    const run = this.stateUpdateQueue.then(() => this.applyStateUpdates(updates));
    this.stateUpdateQueue = run.then(
      () => void 0,
      () => void 0
    );
    return run;
  }
  async updateState(updater) {
    const run = this.stateUpdateQueue.then(async () => {
      const update = await updater(this.getState());
      if (update.updates && Object.keys(update.updates).length > 0) {
        await this.applyStateUpdates(update.updates);
      }
      for (const event of update.events ?? []) {
        this.emit(event);
      }
      return update.result;
    });
    this.stateUpdateQueue = run.then(
      () => void 0,
      () => void 0
    );
    return run;
  }
  getSchemaDefaults() {
    if (!this.stateSchema) return {};
    const defaults = {};
    try {
      const jsonSchema = this.stateSchema["~standard"].jsonSchema.output({ target: "draft-07" });
      if (jsonSchema?.properties) {
        for (const [key, prop] of Object.entries(jsonSchema.properties)) {
          if (prop.default !== void 0) {
            defaults[key] = prop.default;
          }
        }
      }
    } catch {
    }
    return defaults;
  }
  // ===========================================================================
  // Mode Management
  // ===========================================================================
  listModes() {
    return this.config.modes;
  }
  getCurrentModeId() {
    return this.currentModeId;
  }
  getCurrentMode() {
    const mode = this.config.modes.find((m) => m.id === this.currentModeId);
    if (!mode) {
      throw new Error(`Mode not found: ${this.currentModeId}`);
    }
    return mode;
  }
  /**
   * Switch to a different mode.
   * Aborts any in-progress generation and switches to the mode's default model.
   */
  async switchMode({ modeId }) {
    const mode = this.config.modes.find((m) => m.id === modeId);
    if (!mode) {
      throw new Error(`Mode not found: ${modeId}`);
    }
    this.abort();
    const currentModelId = this.getCurrentModelId();
    const previousModeId = this.currentModeId;
    const version = ++this.switchModeVersion;
    this.currentModeId = modeId;
    this.emit({ type: "mode_changed", modeId, previousModeId });
    if (currentModelId) {
      await this.setThreadSetting({ key: `modeModelId_${previousModeId}`, value: currentModelId });
    }
    if (this.switchModeVersion !== version) return;
    await this.setThreadSetting({ key: "currentModeId", value: modeId });
    if (this.switchModeVersion !== version) return;
    const modeModelId = await this.loadModeModelId(modeId);
    if (this.switchModeVersion !== version) return;
    if (modeModelId) {
      void this.setState({ currentModelId: modeModelId });
      this.emit({ type: "model_changed", modelId: modeModelId });
    }
  }
  /**
   * Load the stored model ID for a specific mode.
   * Falls back to: thread metadata -> mode's defaultModelId -> current model.
   */
  async loadModeModelId(modeId) {
    if (this.currentThreadId && this.config.storage) {
      try {
        const memoryStorage = await this.getMemoryStorage();
        const thread = await memoryStorage.getThreadById({ threadId: this.currentThreadId });
        const meta = thread?.metadata;
        const stored = meta?.[`modeModelId_${modeId}`];
        if (stored) return stored;
      } catch {
      }
    }
    const mode = this.config.modes.find((m) => m.id === modeId);
    if (mode?.defaultModelId) return mode.defaultModelId;
    return null;
  }
  propagateRuntimeServicesToAgent(agent) {
    const alreadyHasMastra = !!agent.getMastraInstance();
    const workspaceForAgents = this.workspaceFn ?? this.workspace;
    const browserForAgents = this.browserFn ?? this.browser;
    if (this.config.memory && !agent.hasOwnMemory()) {
      agent.__setMemory(this.config.memory);
    }
    if (workspaceForAgents && !agent.hasOwnWorkspace()) {
      agent.__setWorkspace(workspaceForAgents);
    }
    if (browserForAgents && !agent.hasOwnBrowser()) {
      agent.setBrowser(browserForAgents);
    }
    if (this.config.pubsub && !agent.hasOwnPubSub()) {
      agent.__setPubSub(this.config.pubsub);
    }
    if (this.#internalMastra && !alreadyHasMastra) {
      this.#internalMastra.addAgent(agent);
    }
    return agent;
  }
  /**
   * Get the agent for the current mode.
   */
  getCurrentAgent() {
    const mode = this.getCurrentMode();
    const agent = typeof mode.agent === "function" ? mode.agent(this.state) : mode.agent;
    return this.propagateRuntimeServicesToAgent(agent);
  }
  /**
   * Get a short display name from the current model ID.
   */
  getModelName() {
    const modelId = this.getCurrentModelId();
    if (!modelId || modelId === "unknown") return modelId || "unknown";
    const parts = modelId.split("/");
    return parts[parts.length - 1] || modelId;
  }
  /**
   * Get the full model ID (e.g., "anthropic/claude-sonnet-4").
   */
  getFullModelId() {
    return this.getCurrentModelId();
  }
  /**
   * Switch to a different model at runtime.
   */
  async switchModel({
    modelId,
    scope = "thread",
    modeId
  }) {
    const targetModeId = modeId ?? this.currentModeId;
    if (targetModeId === this.currentModeId) {
      void this.setState({ currentModelId: modelId });
    }
    if (scope === "thread") {
      await this.setThreadSetting({ key: `modeModelId_${targetModeId}`, value: modelId });
    }
    try {
      await Promise.resolve(this.config.modelUseCountTracker?.(modelId));
    } catch (error) {
      console.error("Failed to track model usage count", error);
    }
    this.emit({ type: "model_changed", modelId, scope, modeId: targetModeId });
  }
  getCurrentModelId() {
    const state = this.getState();
    return state.currentModelId ?? "";
  }
  hasModelSelected() {
    return this.getCurrentModelId() !== "";
  }
  /**
   * Check if the current model's provider has authentication configured.
   * Uses the provider registry's `apiKeyEnvVar` and the optional `modelAuthChecker` hook.
   */
  async getCurrentModelAuthStatus() {
    const modelId = this.getCurrentModelId();
    try {
      const availableModels = await this.listAvailableModels();
      const currentModel = availableModels.find((model) => model.id === modelId);
      if (currentModel) {
        if (currentModel.hasApiKey) {
          return { hasAuth: true };
        }
        return { hasAuth: false, apiKeyEnvVar: currentModel.apiKeyEnvVar };
      }
    } catch {
    }
    const provider = modelId.split("/")[0];
    if (!provider) return { hasAuth: true };
    if (this.config.modelAuthChecker) {
      const result = this.config.modelAuthChecker(provider);
      if (result === true) return { hasAuth: true };
      if (result === false) {
        const apiKeyEnvVar = await this.getProviderApiKeyEnvVar(provider);
        return { hasAuth: false, apiKeyEnvVar };
      }
    }
    try {
      const { PROVIDER_REGISTRY } = await import('../provider-registry-TI2S24TW.cjs');
      const registry = PROVIDER_REGISTRY;
      const providerConfig = registry[provider];
      const envVars = providerConfig?.apiKeyEnvVar;
      const apiKeyEnvVar = Array.isArray(envVars) ? envVars[0] : envVars;
      if (apiKeyEnvVar && process.env[apiKeyEnvVar]) {
        return { hasAuth: true };
      }
      return { hasAuth: false, apiKeyEnvVar: apiKeyEnvVar || void 0 };
    } catch {
      return { hasAuth: true };
    }
  }
  /**
   * Get all available models from the provider registry with auth status.
   * Uses the optional `modelAuthChecker`, `modelUseCountProvider`, and
   * `customModelCatalogProvider` hooks.
   */
  async listAvailableModels() {
    const now = Date.now();
    if (this.availableModelsCache && now - this.availableModelsCacheTime < 1e4) {
      return this.availableModelsCache;
    }
    try {
      const { PROVIDER_REGISTRY } = await import('../provider-registry-TI2S24TW.cjs');
      if (!PROVIDER_REGISTRY) return [];
      const registry = PROVIDER_REGISTRY;
      const providers = Object.keys(registry);
      const useCounts = this.config.modelUseCountProvider?.() ?? {};
      const modelsById = /* @__PURE__ */ new Map();
      const upsertModel = (model) => {
        if (!model.id || !model.provider || !model.modelName) return;
        modelsById.set(model.id, {
          ...model,
          useCount: useCounts[model.id] ?? 0
        });
      };
      for (const provider of providers) {
        const providerConfig = registry[provider];
        const envVars = providerConfig?.apiKeyEnvVar;
        const apiKeyEnvVar = Array.isArray(envVars) ? envVars[0] : envVars;
        const hasEnvKey = apiKeyEnvVar ? !!process.env[apiKeyEnvVar] : false;
        let hasApiKey = hasEnvKey;
        if (!hasApiKey && this.config.modelAuthChecker) {
          const customAuth = this.config.modelAuthChecker(provider);
          if (customAuth === true) hasApiKey = true;
        }
        if (providerConfig?.models && Array.isArray(providerConfig.models)) {
          for (const modelName of providerConfig.models) {
            upsertModel({
              id: `${provider}/${modelName}`,
              provider,
              modelName,
              hasApiKey,
              apiKeyEnvVar: apiKeyEnvVar || void 0
            });
          }
        }
      }
      if (this.config.customModelCatalogProvider) {
        try {
          const customModels = await Promise.resolve(this.config.customModelCatalogProvider());
          for (const model of customModels) {
            upsertModel({
              id: model.id,
              provider: model.provider,
              modelName: model.modelName,
              hasApiKey: model.hasApiKey,
              apiKeyEnvVar: model.apiKeyEnvVar
            });
          }
        } catch (error) {
          console.warn("Failed to load custom available models:", error);
        }
      }
      const result = [...modelsById.values()];
      this.availableModelsCache = result;
      this.availableModelsCacheTime = Date.now();
      return result;
    } catch (error) {
      console.warn("Failed to load available models:", error);
      return [];
    }
  }
  invalidateAvailableModelsCache() {
    this.availableModelsCache = null;
    this.availableModelsCacheTime = 0;
  }
  async getProviderApiKeyEnvVar(provider) {
    try {
      const { PROVIDER_REGISTRY } = await import('../provider-registry-TI2S24TW.cjs');
      const registry = PROVIDER_REGISTRY;
      const envVars = registry[provider]?.apiKeyEnvVar;
      return Array.isArray(envVars) ? envVars[0] : envVars;
    } catch {
      return void 0;
    }
  }
  // ===========================================================================
  // Thread Management
  // ===========================================================================
  getCurrentThreadId() {
    return this.currentThreadId;
  }
  getResourceId() {
    return this.resourceId;
  }
  async getResolvedMemory() {
    if (!this.config.memory) return null;
    return this.resolveMemory();
  }
  setResourceId({ resourceId }) {
    this.cleanupAgentThreadSubscription();
    this.resourceId = resourceId;
    this.currentThreadId = null;
  }
  getDefaultResourceId() {
    return this.defaultResourceId;
  }
  async getKnownResourceIds() {
    const threads = await this.listThreads({ allResources: true });
    const ids = new Set(threads.map((t) => t.resourceId));
    return [...ids].sort();
  }
  async createThread({ title } = {}) {
    this.cleanupAgentThreadSubscription();
    const now = /* @__PURE__ */ new Date();
    const thread = {
      id: this.generateId(),
      resourceId: this.resourceId,
      title: title || "",
      createdAt: now,
      updatedAt: now
    };
    const currentStateModel = this.state.currentModelId;
    const currentMode = this.getCurrentMode();
    const modelId = currentStateModel || currentMode.defaultModelId;
    const metadata = {};
    if (modelId) {
      metadata.currentModelId = modelId;
      metadata[`modeModelId_${this.currentModeId}`] = modelId;
    }
    const projectPath = this.state.projectPath;
    if (projectPath) {
      metadata.projectPath = projectPath;
    }
    const oldThreadId = this.currentThreadId;
    if (this.config.threadLock) {
      try {
        await this.config.threadLock.acquire(thread.id);
      } catch (err) {
        if (oldThreadId) {
          try {
            await this.config.threadLock.acquire(oldThreadId);
          } catch {
          }
        }
        throw err;
      }
      if (oldThreadId) {
        await this.config.threadLock.release(oldThreadId);
      }
    }
    if (this.config.storage) {
      const memoryStorage = await this.getMemoryStorage();
      try {
        await memoryStorage.saveThread({
          thread: {
            id: thread.id,
            resourceId: thread.resourceId,
            title: thread.title,
            createdAt: thread.createdAt,
            updatedAt: thread.updatedAt,
            metadata: Object.keys(metadata).length > 0 ? metadata : void 0
          }
        });
      } catch (err) {
        let reacquired = false;
        if (this.config.threadLock) {
          try {
            await this.config.threadLock.release(thread.id);
          } catch {
          }
          if (oldThreadId) {
            try {
              await this.config.threadLock.acquire(oldThreadId);
              reacquired = true;
            } catch {
            }
          }
        }
        this.currentThreadId = reacquired ? oldThreadId : null;
        throw err;
      }
    }
    this.currentThreadId = thread.id;
    if (modelId && !currentStateModel) {
      void this.setState({ currentModelId: modelId });
    }
    this.tokenUsage = createEmptyTokenUsage();
    this.emit({ type: "thread_created", thread });
    await this.ensureCurrentAgentThreadSubscription();
    return thread;
  }
  /**
   * Returns a memory accessor with thread and message management methods.
   */
  get memory() {
    return {
      createThread: this.createThread.bind(this),
      switchThread: this.switchThread.bind(this),
      listThreads: this.listThreads.bind(this),
      renameThread: this.renameThread.bind(this),
      deleteThread: this.deleteThread.bind(this)
    };
  }
  async deleteThread({ threadId }) {
    if (!this.config.storage) return;
    const memoryStorage = await this.getMemoryStorage();
    const thread = await memoryStorage.getThreadById({ threadId });
    if (!thread) {
      throw new Error(`Thread not found: ${threadId}`);
    }
    const isDeletingCurrentThread = this.currentThreadId === threadId;
    await memoryStorage.deleteThread({ threadId });
    if (isDeletingCurrentThread) {
      try {
        await this.config.threadLock?.release(threadId);
      } catch {
      }
      this.cleanupAgentThreadSubscription();
      this.currentThreadId = null;
      this.tokenUsage = createEmptyTokenUsage();
    }
    this.emit({ type: "thread_deleted", threadId });
  }
  async renameThread({ title }) {
    if (!this.currentThreadId || !this.config.storage) return;
    const memoryStorage = await this.getMemoryStorage();
    const thread = await memoryStorage.getThreadById({ threadId: this.currentThreadId });
    if (thread) {
      await memoryStorage.saveThread({
        thread: { ...thread, title, updatedAt: /* @__PURE__ */ new Date() }
      });
    }
  }
  async cloneThread({
    sourceThreadId,
    title,
    resourceId
  } = {}) {
    const sourceId = sourceThreadId ?? this.currentThreadId;
    if (!sourceId) {
      throw new Error("No source thread to clone");
    }
    if (!this.config.memory) {
      throw new Error("Memory is not configured on this Harness");
    }
    const memory = await this.resolveMemory();
    const result = await memory.cloneThread({
      sourceThreadId: sourceId,
      resourceId: resourceId ?? this.resourceId,
      title
    });
    const clonedThread = {
      id: result.thread.id,
      resourceId: result.thread.resourceId,
      title: result.thread.title ?? "Cloned Thread",
      createdAt: result.thread.createdAt,
      updatedAt: result.thread.updatedAt,
      metadata: result.thread.metadata
    };
    const oldThreadId = this.currentThreadId;
    if (this.config.threadLock) {
      try {
        await this.config.threadLock.acquire(clonedThread.id);
      } catch (err) {
        if (oldThreadId) {
          try {
            await this.config.threadLock.acquire(oldThreadId);
          } catch {
          }
        }
        throw err;
      }
      if (oldThreadId) {
        await this.config.threadLock.release(oldThreadId);
      }
    }
    this.cleanupAgentThreadSubscription();
    this.currentThreadId = clonedThread.id;
    await this.loadThreadMetadata();
    this.tokenUsage = createEmptyTokenUsage();
    this.emit({ type: "thread_created", thread: clonedThread });
    await this.ensureCurrentAgentThreadSubscription();
    return clonedThread;
  }
  async switchThread({ threadId }) {
    this.abort();
    this.cleanupAgentThreadSubscription();
    await this.config.threadLock?.acquire(threadId);
    const previousThreadId = this.currentThreadId;
    if (previousThreadId) {
      await this.config.threadLock?.release(previousThreadId);
    }
    if (this.config.storage) {
      const memoryStorage = await this.getMemoryStorage();
      const thread = await memoryStorage.getThreadById({ threadId });
      if (!thread) {
        throw new Error(`Thread not found: ${threadId}`);
      }
    }
    this.currentThreadId = threadId;
    await this.loadThreadMetadata();
    this.emit({ type: "thread_changed", threadId, previousThreadId });
    await this.ensureCurrentAgentThreadSubscription();
  }
  async listThreads(options) {
    if (!this.config.storage) return [];
    const memoryStorage = await this.getMemoryStorage();
    const filter = options?.allResources ? void 0 : { resourceId: this.resourceId };
    const result = await memoryStorage.listThreads({ filter, perPage: false });
    const threads = options?.includeForkedSubagents ? result.threads : result.threads.filter((thread) => {
      const metadata = thread.metadata;
      return metadata?.forkedSubagent !== true;
    });
    return threads.map((thread) => ({
      id: thread.id,
      resourceId: thread.resourceId,
      title: thread.title,
      createdAt: thread.createdAt,
      updatedAt: thread.updatedAt,
      metadata: thread.metadata
    }));
  }
  async setThreadSetting({ key, value }) {
    if (!this.currentThreadId || !this.config.storage) return;
    try {
      const memoryStorage = await this.getMemoryStorage();
      const thread = await memoryStorage.getThreadById({ threadId: this.currentThreadId });
      if (thread) {
        await memoryStorage.saveThread({
          thread: {
            ...thread,
            metadata: { ...thread.metadata, [key]: value },
            updatedAt: /* @__PURE__ */ new Date()
          }
        });
      }
    } catch {
    }
  }
  async deleteThreadSetting({ key }) {
    if (!this.currentThreadId || !this.config.storage) return;
    try {
      const memoryStorage = await this.getMemoryStorage();
      const thread = await memoryStorage.getThreadById({ threadId: this.currentThreadId });
      if (thread && thread.metadata) {
        const metadata = { ...thread.metadata };
        delete metadata[key];
        await memoryStorage.saveThread({
          thread: {
            ...thread,
            metadata: Object.keys(metadata).length > 0 ? metadata : void 0,
            updatedAt: /* @__PURE__ */ new Date()
          }
        });
      }
    } catch {
    }
  }
  async loadThreadMetadata() {
    if (!this.currentThreadId || !this.config.storage) {
      this.tokenUsage = createEmptyTokenUsage();
      return;
    }
    try {
      const memoryStorage = await this.getMemoryStorage();
      const thread = await memoryStorage.getThreadById({ threadId: this.currentThreadId });
      const savedUsage = thread?.metadata?.tokenUsage;
      if (savedUsage) {
        this.tokenUsage = {
          ...createEmptyTokenUsage(),
          ...savedUsage,
          promptTokens: savedUsage.promptTokens ?? 0,
          completionTokens: savedUsage.completionTokens ?? 0,
          totalTokens: savedUsage.totalTokens ?? 0,
          cachedInputTokens: savedUsage.cachedInputTokens ?? 0,
          cacheCreationInputTokens: savedUsage.cacheCreationInputTokens ?? 0
        };
      } else {
        this.tokenUsage = createEmptyTokenUsage();
      }
      const meta = thread?.metadata;
      const updates = {};
      let previousModeIdForEmit;
      if (meta?.currentModeId) {
        const savedModeId = meta.currentModeId;
        const modeExists = this.config.modes.some((m) => m.id === savedModeId);
        if (modeExists && savedModeId !== this.currentModeId) {
          previousModeIdForEmit = this.currentModeId;
          this.currentModeId = savedModeId;
        }
      }
      const modeModelKey = `modeModelId_${this.currentModeId}`;
      if (meta?.[modeModelKey]) {
        updates.currentModelId = meta[modeModelKey];
      } else {
        const currentMode = this.config.modes.find((m) => m.id === this.currentModeId);
        if (currentMode?.defaultModelId) {
          updates.currentModelId = currentMode.defaultModelId;
        } else if (meta?.currentModelId) {
          updates.currentModelId = meta.currentModelId;
        }
      }
      if (previousModeIdForEmit !== void 0) {
        this.emit({
          type: "mode_changed",
          modeId: this.currentModeId,
          previousModeId: previousModeIdForEmit
        });
      }
      if (meta?.observerModelId) {
        updates.observerModelId = meta.observerModelId;
      }
      if (meta?.reflectorModelId) {
        updates.reflectorModelId = meta.reflectorModelId;
      }
      const hasObservationThreshold = typeof meta?.observationThreshold === "number";
      const hasReflectionThreshold = typeof meta?.reflectionThreshold === "number";
      if (hasObservationThreshold) {
        updates.observationThreshold = meta.observationThreshold;
      }
      if (hasReflectionThreshold) {
        updates.reflectionThreshold = meta.reflectionThreshold;
      }
      if (Object.keys(updates).length > 0) {
        await this.setState(updates);
      }
      if (!hasObservationThreshold) {
        const observationThreshold = this.getObservationThreshold();
        if (observationThreshold !== void 0) {
          await this.setThreadSetting({ key: "observationThreshold", value: observationThreshold });
        }
      }
      if (!hasReflectionThreshold) {
        const reflectionThreshold = this.getReflectionThreshold();
        if (reflectionThreshold !== void 0) {
          await this.setThreadSetting({ key: "reflectionThreshold", value: reflectionThreshold });
        }
      }
    } catch {
      this.tokenUsage = createEmptyTokenUsage();
    }
  }
  // ===========================================================================
  // Observational Memory
  // ===========================================================================
  /**
   * Load observational memory progress for the current thread.
   * Reads the OM record and recent messages to reconstruct status,
   * then emits an `om_status` event for the UI.
   */
  async loadOMProgress() {
    if (!this.currentThreadId) return;
    try {
      const memoryStorage = await this.getMemoryStorage();
      const record = await memoryStorage.getObservationalMemory(this.currentThreadId, this.resourceId);
      if (!record) return;
      const config = record.config;
      const getThreshold = (val, fallback) => {
        if (!val) return fallback;
        if (typeof val === "number") return val;
        return val.max;
      };
      let observationThreshold = getThreshold(config?.observationThreshold, 3e4);
      let reflectionThreshold = getThreshold(config?.reflectionThreshold, 4e4);
      let messageTokens = record.pendingMessageTokens ?? 0;
      let observationTokens = record.observationTokenCount ?? 0;
      let bufferedObs = {
        status: "idle",
        chunks: 0,
        messageTokens: 0,
        projectedMessageRemoval: 0,
        observationTokens: 0
      };
      let bufferedRef = {
        status: "idle",
        inputObservationTokens: 0,
        observationTokens: 0
      };
      let generationCount = 0;
      let stepNumber = 0;
      const messagesResult = await memoryStorage.listMessages({
        threadId: this.currentThreadId,
        perPage: 70,
        page: 0,
        orderBy: { field: "createdAt", direction: "DESC" }
      });
      const messages = messagesResult.messages;
      let foundStatus = false;
      for (const msg of messages) {
        if (msg.role !== "assistant") continue;
        const content = msg.content;
        if (typeof content === "string" || !content?.parts) continue;
        for (let i = content.parts.length - 1; i >= 0; i--) {
          const part = content.parts[i];
          if (part.type === "data-om-status" && part.data?.windows) {
            const w = part.data.windows;
            messageTokens = w.active?.messages?.tokens ?? messageTokens;
            observationTokens = w.active?.observations?.tokens ?? observationTokens;
            const msgThresh = w.active?.messages?.threshold;
            const obsThresh = w.active?.observations?.threshold;
            if (msgThresh) observationThreshold = msgThresh;
            if (obsThresh) reflectionThreshold = obsThresh;
            const bo = w.buffered?.observations;
            if (bo) {
              bufferedObs = {
                status: bo.status ?? "idle",
                chunks: bo.chunks ?? 0,
                messageTokens: bo.messageTokens ?? 0,
                projectedMessageRemoval: bo.projectedMessageRemoval ?? 0,
                observationTokens: bo.observationTokens ?? 0
              };
            }
            const br = w.buffered?.reflection;
            if (br) {
              bufferedRef = {
                status: br.status ?? "idle",
                inputObservationTokens: br.inputObservationTokens ?? 0,
                observationTokens: br.observationTokens ?? 0
              };
            }
            generationCount = part.data.generationCount ?? 0;
            stepNumber = part.data.stepNumber ?? 0;
            foundStatus = true;
            break;
          }
        }
        if (foundStatus) break;
      }
      this.emit({
        type: "om_status",
        windows: {
          active: {
            messages: { tokens: messageTokens, threshold: observationThreshold },
            observations: { tokens: observationTokens, threshold: reflectionThreshold }
          },
          buffered: { observations: bufferedObs, reflection: bufferedRef }
        },
        recordId: record.id ?? "",
        threadId: this.currentThreadId,
        stepNumber,
        generationCount
      });
    } catch {
    }
  }
  async getObservationalMemoryRecord() {
    if (!this.currentThreadId) return null;
    try {
      const memoryStorage = await this.getMemoryStorage();
      return await memoryStorage.getObservationalMemory(this.currentThreadId, this.resourceId);
    } catch {
      return null;
    }
  }
  /**
   * Returns the observer model ID from state, falling back to omConfig defaults.
   */
  getObserverModelId() {
    return this.state.observerModelId ?? this.config.omConfig?.defaultObserverModelId;
  }
  /**
   * Returns the reflector model ID from state, falling back to omConfig defaults.
   */
  getReflectorModelId() {
    return this.state.reflectorModelId ?? this.config.omConfig?.defaultReflectorModelId;
  }
  /**
   * Returns the observation threshold from state, falling back to omConfig defaults.
   */
  getObservationThreshold() {
    return this.state.observationThreshold ?? this.config.omConfig?.defaultObservationThreshold;
  }
  /**
   * Returns the reflection threshold from state, falling back to omConfig defaults.
   */
  getReflectionThreshold() {
    return this.state.reflectionThreshold ?? this.config.omConfig?.defaultReflectionThreshold;
  }
  /**
   * Resolves the observer model ID to a language model instance via `resolveModel`.
   */
  getResolvedObserverModel() {
    const modelId = this.getObserverModelId();
    if (!modelId || !this.config.resolveModel) return void 0;
    return this.config.resolveModel(modelId);
  }
  /**
   * Resolves the reflector model ID to a language model instance via `resolveModel`.
   */
  getResolvedReflectorModel() {
    const modelId = this.getReflectorModelId();
    if (!modelId || !this.config.resolveModel) return void 0;
    return this.config.resolveModel(modelId);
  }
  /**
   * Switch the Observer model.
   */
  async switchObserverModel({ modelId }) {
    void this.setState({ observerModelId: modelId });
    await this.setThreadSetting({ key: "observerModelId", value: modelId });
    this.emit({ type: "om_model_changed", role: "observer", modelId });
  }
  /**
   * Switch the Reflector model.
   */
  async switchReflectorModel({ modelId }) {
    void this.setState({ reflectorModelId: modelId });
    await this.setThreadSetting({ key: "reflectorModelId", value: modelId });
    this.emit({ type: "om_model_changed", role: "reflector", modelId });
  }
  // ===========================================================================
  // Subagent Model Management
  // ===========================================================================
  getSubagentModelId({ agentType } = {}) {
    const state = this.state;
    if (agentType) {
      const perType = state[`subagentModelId_${agentType}`];
      if (typeof perType === "string") return perType;
    }
    const global = state.subagentModelId;
    return typeof global === "string" ? global : null;
  }
  async setSubagentModelId({ modelId, agentType }) {
    const key = agentType ? `subagentModelId_${agentType}` : "subagentModelId";
    void this.setState({ [key]: modelId });
    await this.setThreadSetting({ key, value: modelId });
    this.emit({ type: "subagent_model_changed", modelId, scope: "thread", agentType });
  }
  // ===========================================================================
  // Permissions
  // ===========================================================================
  grantSessionCategory({ category }) {
    this.sessionGrantedCategories.add(category);
  }
  grantSessionTool({ toolName }) {
    this.sessionGrantedTools.add(toolName);
  }
  getSessionGrants() {
    return {
      categories: [...this.sessionGrantedCategories],
      tools: [...this.sessionGrantedTools]
    };
  }
  getToolCategory({ toolName }) {
    return this.config.toolCategoryResolver?.(toolName) ?? null;
  }
  setPermissionForCategory({ category, policy }) {
    const rules = this.getPermissionRules();
    rules.categories[category] = policy;
    void this.setState({ permissionRules: rules });
  }
  setPermissionForTool({ toolName, policy }) {
    const rules = this.getPermissionRules();
    rules.tools[toolName] = policy;
    void this.setState({ permissionRules: rules });
  }
  getPermissionRules() {
    const state = this.state;
    const rules = state.permissionRules;
    return rules ?? { categories: {}, tools: {} };
  }
  /**
   * Resolve whether a tool call should be auto-approved, denied, or asked.
   * Resolution chain: per-tool deny → yolo → per-tool policy → session tool grant →
   * session category grant → category policy → "ask"
   */
  resolveToolApproval(toolName) {
    const state = this.state;
    const rules = this.getPermissionRules();
    const toolPolicy = rules.tools[toolName];
    if (toolPolicy === "deny") return "deny";
    if (state.yolo === true) return "allow";
    if (toolPolicy) return toolPolicy;
    if (this.sessionGrantedTools.has(toolName)) return "allow";
    const category = this.getToolCategory({ toolName });
    if (category) {
      if (this.sessionGrantedCategories.has(category)) return "allow";
      const categoryPolicy = rules.categories[category];
      if (categoryPolicy) return categoryPolicy;
    }
    return "ask";
  }
  // ===========================================================================
  // Message Handling
  // ===========================================================================
  cleanupAgentThreadSubscription() {
    this.agentThreadSubscription?.abort();
    this.agentThreadSubscription?.unsubscribe();
    this.agentThreadSubscription = null;
    this.agentThreadSubscriptionKey = null;
    this.currentRunId = null;
    this.currentTraceId = null;
    this.abortController = null;
    this.abortRequested = false;
  }
  getAgentThreadSubscriptionKey(agent, threadId) {
    return `${agent.id}:${this.resourceId}:${threadId}`;
  }
  async ensureAgentThreadSubscription(agent, threadId) {
    const key = this.getAgentThreadSubscriptionKey(agent, threadId);
    if (this.agentThreadSubscriptionKey === key && this.agentThreadSubscription) return;
    this.cleanupAgentThreadSubscription();
    const subscription = await agent.subscribeToThread({ resourceId: this.resourceId, threadId });
    this.agentThreadSubscription = subscription;
    this.agentThreadSubscriptionKey = key;
    void this.processSubscribedThreadStream(subscription);
  }
  async ensureCurrentAgentThreadSubscription() {
    if (!this.currentThreadId) return;
    await this.ensureAgentThreadSubscription(this.getCurrentAgent(), this.currentThreadId);
  }
  createMessageInput({
    content,
    files
  }) {
    if (!files?.length) return content;
    const fileParts = files.map((f) => {
      const isText = f.mediaType.startsWith("text/") || f.mediaType === "application/json";
      if (isText) {
        let textContent = f.data;
        const base64Match = f.data.match(/^data:[^;]*;base64,(.*)$/);
        if (base64Match) {
          try {
            textContent = Buffer.from(base64Match[1], "base64").toString("utf-8");
          } catch {
          }
        }
        const label = f.filename ? `[File: ${f.filename}]` : "[Attached file]";
        const maxBacktickRun = Math.max(0, ...Array.from(textContent.matchAll(/`+/g), (match) => match[0].length));
        const fence = "`".repeat(Math.max(3, maxBacktickRun + 1));
        return { type: "text", text: `${label}
${fence}
${textContent}
${fence}` };
      }
      return {
        type: "file",
        data: f.data,
        mediaType: f.mediaType,
        ...f.filename ? { filename: f.filename } : {}
      };
    });
    return [{ type: "text", text: content }, ...fileParts];
  }
  async buildAgentMessageStreamOptions({
    requestContext: requestContextInput,
    tracingContext,
    tracingOptions
  }) {
    if (!this.currentThreadId) {
      throw new Error("Cannot build stream options without a current thread");
    }
    this.abortRequested = false;
    this.abortController ??= new AbortController();
    const requestContext = await this.buildRequestContext(requestContextInput);
    const isYolo = this.state.yolo === true;
    const streamOptions = {
      memory: { thread: this.currentThreadId, resource: this.resourceId },
      abortSignal: this.abortController.signal,
      requestContext,
      maxSteps: 1e3,
      savePerStep: false,
      requireToolApproval: !isYolo,
      modelSettings: { temperature: 1 },
      ...tracingContext && { tracingContext },
      ...tracingOptions && { tracingOptions }
    };
    streamOptions.toolsets = await this.buildToolsets(requestContext);
    return streamOptions;
  }
  async drainFollowUpQueue(options) {
    if (this.followUpQueue.length === 0) return;
    const next = this.followUpQueue.shift();
    try {
      if (this.agentThreadSubscription && this.currentThreadId) {
        const agent = this.getCurrentAgent();
        const streamOptions = await this.buildAgentMessageStreamOptions({
          requestContext: next.requestContext,
          tracingContext: options?.tracingContext,
          tracingOptions: options?.tracingOptions
        });
        const result = agent.queueMessage(this.createMessageInput({ content: next.content }), {
          resourceId: this.resourceId,
          threadId: this.currentThreadId,
          ifIdle: { streamOptions }
        });
        this.emit({ type: "follow_up_queued", count: this.followUpQueue.length, runId: result.runId });
      } else {
        this.emit({ type: "follow_up_queued", count: this.followUpQueue.length });
        await this.sendMessage({
          content: next.content,
          requestContext: next.requestContext,
          tracingContext: options?.tracingContext,
          tracingOptions: options?.tracingOptions
        });
      }
    } catch (error) {
      this.followUpQueue.unshift(next);
      this.emit({ type: "follow_up_queued", count: this.followUpQueue.length });
      throw error;
    }
  }
  isActiveAgentThreadSubscription(subscription) {
    return this.agentThreadSubscription === subscription;
  }
  async finishSubscribedStreamRun({
    suspended,
    error,
    aborted
  }) {
    const reason = error ? "error" : suspended ? "suspended" : aborted || this.abortRequested ? "aborted" : "complete";
    this.emit({ type: "agent_end", reason });
    this.currentRunId = null;
    this.currentTraceId = null;
    this.abortController = null;
    this.abortRequested = false;
    await this.drainFollowUpQueue();
  }
  async handleSubscribedStreamError(error) {
    if (error instanceof Error && error.name === "AbortError") {
      this.emit({ type: "agent_end", reason: "aborted" });
    } else {
      this.emit({ type: "error", error: error instanceof Error ? error : new Error(String(error)) });
      this.emit({ type: "agent_end", reason: "error" });
    }
    this.agentThreadSubscription?.unsubscribe();
    this.agentThreadSubscription = null;
    this.agentThreadSubscriptionKey = null;
    this.currentRunId = null;
    this.currentTraceId = null;
    this.abortController = null;
    this.abortRequested = false;
    await this.drainFollowUpQueue();
  }
  async processSubscribedThreadStream(subscription) {
    const requestContext = await this.buildRequestContext();
    let currentRun;
    let lastFinishedRunId = null;
    try {
      for await (const chunk of subscription.stream) {
        if (!this.isActiveAgentThreadSubscription(subscription)) {
          subscription.unsubscribe();
          break;
        }
        const chunkRunId = "runId" in chunk ? chunk.runId : null;
        if (lastFinishedRunId && chunkRunId === lastFinishedRunId) {
          continue;
        }
        if (!currentRun) {
          currentRun = this.createStreamState();
          this.currentOperationId += 1;
          this.abortController ??= new AbortController();
          this.currentRunId = subscription.activeRunId() ?? ("runId" in chunk ? chunk.runId : null);
          this.currentTraceId = null;
          this.emit({ type: "agent_start" });
        }
        if (chunk.type === "start") {
          continue;
        }
        try {
          const streamResult = await this.processStreamChunk(currentRun, chunk, requestContext);
          if (streamResult || chunk.type === "finish" || chunk.type === "error" || chunk.type === "abort" || chunk.type === "tool-call-suspended") {
            const finishedRunId = chunkRunId ?? this.currentRunId;
            await this.finishSubscribedStreamRun({
              suspended: chunk.type === "tool-call-suspended" || (streamResult ?? this.finishStreamState(currentRun)).suspended || void 0,
              error: chunk.type === "error",
              aborted: chunk.type === "abort"
            });
            lastFinishedRunId = finishedRunId;
            currentRun = void 0;
          }
        } catch (error) {
          await this.handleSubscribedStreamError(error);
          currentRun = void 0;
        }
      }
    } catch (error) {
      if (this.isActiveAgentThreadSubscription(subscription)) {
        await this.handleSubscribedStreamError(error);
      }
    }
  }
  /**
   * Send a signal to the current agent/thread.
   */
  sendSignal(input) {
    const { tracingContext, tracingOptions, requestContext: requestContextInput } = "content" in input ? input : {};
    const ifActive = "content" in input ? input.ifActive : void 0;
    const ifIdle = "content" in input ? input.ifIdle : void 0;
    const signal = chunkDCCAHUCG_cjs.createSignal(
      "content" in input ? { type: "user", tagName: "user", contents: input.content } : input
    );
    const accepted = Promise.resolve().then(async () => {
      if (!this.currentThreadId) {
        const thread = await this.createThread();
        this.currentThreadId = thread.id;
      }
      const agent = this.getCurrentAgent();
      await this.ensureAgentThreadSubscription(agent, this.currentThreadId);
      if (this.currentRunId && this.agentThreadSubscription?.activeRunId()) {
        const result2 = agent.sendSignal(signal, {
          resourceId: this.resourceId,
          threadId: this.currentThreadId,
          ifActive,
          ifIdle
        });
        return { accepted: result2.accepted, runId: result2.runId };
      }
      const streamOptions = await this.buildAgentMessageStreamOptions({
        requestContext: requestContextInput,
        tracingContext,
        tracingOptions
      });
      const result = agent.sendSignal(signal, {
        resourceId: this.resourceId,
        threadId: this.currentThreadId,
        ifActive,
        ifIdle: { ...ifIdle, streamOptions }
      });
      return { accepted: result.accepted, runId: result.runId };
    });
    return { id: signal.id, type: signal.type, accepted };
  }
  /**
   * Send a message to the current agent.
   * Streams the response and emits events.
   */
  async sendMessage({
    content,
    files,
    tracingContext,
    tracingOptions,
    requestContext: requestContextInput
  }) {
    const messageInput = this.createMessageInput({ content, files });
    const wasActive = this.isCurrentThreadStreamActive();
    let emittedAgentEnd = false;
    const unsubscribeAgentEnd = wasActive ? void 0 : this.subscribe((event) => {
      if (event.type === "agent_end") emittedAgentEnd = true;
    });
    const signal = this.sendSignal({
      content: messageInput,
      tracingContext,
      tracingOptions,
      requestContext: requestContextInput
    });
    await signal.accepted;
    if (!wasActive) {
      await new Promise((resolve) => setTimeout(resolve, 0));
      await this.waitForCurrentThreadStreamIdle();
      unsubscribeAgentEnd?.();
      if (!emittedAgentEnd && !this.pendingSuspensionRunId) {
        this.emit({ type: "agent_end", reason: "complete" });
      }
    }
    return;
  }
  async listMessages(options) {
    if (!this.currentThreadId) return [];
    return this.listMessagesForThread({ threadId: this.currentThreadId, limit: options?.limit });
  }
  async saveSystemReminderMessage({
    message,
    reminderType,
    role = "user",
    metadata
  }) {
    if (!this.currentThreadId || !this.config.storage) return null;
    const memoryStorage = await this.getMemoryStorage();
    const dbMessage = {
      id: crypto.randomUUID(),
      role,
      threadId: this.currentThreadId,
      resourceId: this.resourceId,
      createdAt: /* @__PURE__ */ new Date(),
      content: {
        format: 2,
        parts: [],
        content: "",
        metadata: {
          systemReminder: {
            type: reminderType,
            message,
            ...metadata
          }
        }
      }
    };
    const result = await memoryStorage.saveMessages({ messages: [dbMessage] });
    const saved = result.messages[0] ?? dbMessage;
    return this.convertToHarnessMessage(saved);
  }
  async listMessagesForThread({ threadId, limit }) {
    if (!this.config.storage) return [];
    const memoryStorage = await this.getMemoryStorage();
    if (limit) {
      const result2 = await memoryStorage.listMessages({
        threadId,
        perPage: limit,
        page: 0,
        orderBy: { field: "createdAt", direction: "DESC" }
      });
      return result2.messages.map((msg) => this.convertToHarnessMessage(msg)).reverse();
    }
    const result = await memoryStorage.listMessages({ threadId, perPage: false });
    return result.messages.map((msg) => this.convertToHarnessMessage(msg));
  }
  async getFirstUserMessageForThread({ threadId }) {
    const messages = await this.getFirstUserMessagesForThreads({ threadIds: [threadId] });
    return messages.get(threadId) ?? null;
  }
  async getFirstUserMessagesForThreads({ threadIds }) {
    if (!this.config.storage || threadIds.length === 0) return /* @__PURE__ */ new Map();
    const memoryStorage = await this.getMemoryStorage();
    const result = await memoryStorage.listMessages({
      threadId: threadIds,
      perPage: false,
      orderBy: { field: "createdAt", direction: "ASC" }
    });
    const firstUserMessages = /* @__PURE__ */ new Map();
    for (const message of result.messages) {
      if (message.role !== "user" || !message.threadId || firstUserMessages.has(message.threadId)) continue;
      firstUserMessages.set(message.threadId, this.convertToHarnessMessage(message));
      if (firstUserMessages.size === threadIds.length) {
        break;
      }
    }
    return firstUserMessages;
  }
  convertToHarnessMessage(msg) {
    const content = [];
    const systemReminder = getRecordValue(msg.content.metadata?.systemReminder);
    if (systemReminder && typeof systemReminder.type === "string") {
      const reminder = toSystemReminderContent({
        ...systemReminder,
        contents: typeof systemReminder.message === "string" ? systemReminder.message : "",
        reminderType: systemReminder.type
      });
      if (reminder) {
        content.push(reminder);
      }
      return {
        id: msg.id,
        role: msg.role === "signal" ? "user" : msg.role,
        content,
        createdAt: msg.createdAt
      };
    }
    if (msg.role === "signal") {
      const signal = chunkDCCAHUCG_cjs.mastraDBMessageToSignal(msg);
      if (signal.type === "user") {
        const signalContent = signalContentsToHarnessContent(signal.contents);
        if (signalContent.length > 0) {
          return {
            id: msg.id,
            role: "user",
            content: signalContent,
            createdAt: msg.createdAt,
            attributes: signal.attributes
          };
        }
      }
      if (signal.type === "reactive" && signal.tagName === "system-reminder") {
        const reminder = toSystemReminderContent({
          type: signal.type,
          contents: typeof signal.contents === "string" ? signal.contents : signal.contents.filter((p) => p.type === "text").map((p) => p.text).join("\n"),
          attributes: signal.attributes ?? msg.content.metadata,
          metadata: signal.metadata
        });
        if (reminder) {
          content.push(reminder);
        }
        return {
          id: msg.id,
          role: "user",
          content,
          createdAt: msg.createdAt
        };
      }
    }
    for (const part of msg.content.parts) {
      switch (part.type) {
        case "text":
          if (part.text) {
            content.push({ type: "text", text: part.text });
          }
          break;
        case "reasoning":
          if (part.reasoning) {
            content.push({ type: "thinking", thinking: part.reasoning });
          }
          break;
        case "tool-invocation":
          if (part.toolInvocation) {
            const inv = part.toolInvocation;
            content.push({ type: "tool_call", id: inv.toolCallId, name: inv.toolName, args: inv.args });
            if (inv.state === "result" && inv.result !== void 0) {
              const partProviderMetadata = part.providerMetadata;
              content.push({
                type: "tool_result",
                id: inv.toolCallId,
                name: inv.toolName,
                result: inv.result,
                isError: inv.isError ?? false,
                ...partProviderMetadata ? { providerMetadata: partProviderMetadata } : {}
              });
            }
          } else if (part.toolCallId && part.toolName) {
            content.push({ type: "tool_call", id: part.toolCallId, name: part.toolName, args: part.args });
          }
          break;
        case "tool-call":
          if (part.toolCallId && part.toolName) {
            content.push({ type: "tool_call", id: part.toolCallId, name: part.toolName, args: part.args });
          }
          break;
        case "tool-result":
          if (part.toolCallId && part.toolName) {
            const resultProviderMetadata = part.providerMetadata;
            content.push({
              type: "tool_result",
              id: part.toolCallId,
              name: part.toolName,
              result: part.result,
              isError: part.isError ?? false,
              ...resultProviderMetadata ? { providerMetadata: resultProviderMetadata } : {}
            });
          }
          break;
        case "data-om-observation-start": {
          const data = part.data ?? {};
          content.push({
            type: "om_observation_start",
            tokensToObserve: data.tokensToObserve ?? 0,
            operationType: data.operationType ?? "observation"
          });
          break;
        }
        case "data-om-observation-end": {
          const data = part.data ?? {};
          content.push({
            type: "om_observation_end",
            tokensObserved: data.tokensObserved ?? 0,
            observationTokens: data.observationTokens ?? 0,
            durationMs: data.durationMs ?? 0,
            operationType: data.operationType ?? "observation",
            observations: data.observations ?? void 0,
            currentTask: data.currentTask ?? void 0,
            suggestedResponse: data.suggestedResponse ?? void 0
          });
          break;
        }
        case "data-om-observation-failed": {
          const data = part.data ?? {};
          content.push({
            type: "om_observation_failed",
            error: data.error ?? "Unknown error",
            tokensAttempted: data.tokensAttempted ?? 0,
            operationType: data.operationType ?? "observation"
          });
          break;
        }
        case "data-user-message": {
          const data = part.data ?? {};
          const message = toUserSignalMessage(data);
          if (message) {
            content.push(...message.content);
          }
          break;
        }
        case "data-system-reminder": {
          const data = part.data ?? {};
          const reminder = toSystemReminderContent(data);
          if (reminder) {
            content.push(reminder);
          }
          break;
        }
        case "file":
          if (typeof part.data !== "string") {
            console.warn("[Harness] Skipping file part with non-string data:", typeof part.data);
            break;
          }
          content.push({
            type: "file",
            data: part.data,
            mediaType: part.mediaType ?? part.mimeType ?? "application/octet-stream",
            ...part.filename ? { filename: part.filename } : {}
          });
          break;
        case "image": {
          const imgData = typeof part.data === "string" ? part.data : typeof part.image === "string" ? part.image : "";
          content.push({
            type: "image",
            data: imgData,
            mimeType: part.mimeType ?? part.mediaType ?? "image/png"
          });
          break;
        }
        case "data-om-thread-update": {
          const data = part.data ?? {};
          if (data.newTitle) {
            content.push({
              type: "om_thread_title_updated",
              threadId: data.threadId ?? "",
              oldTitle: data.oldTitle ?? void 0,
              newTitle: data.newTitle
            });
          }
          break;
        }
      }
    }
    return { id: msg.id, role: msg.role === "signal" ? "user" : msg.role, content, createdAt: msg.createdAt };
  }
  /**
   * Process a stream response (shared between sendMessage and tool approval).
   */
  createStreamState() {
    return {
      currentMessage: {
        id: this.generateId(),
        role: "assistant",
        content: [],
        createdAt: /* @__PURE__ */ new Date()
      },
      isSuspended: false,
      textContentById: /* @__PURE__ */ new Map(),
      thinkingContentById: /* @__PURE__ */ new Map()
    };
  }
  abortForOmFailure({ operationType, stage, error }) {
    this.emit({
      type: "error",
      error: new Error(`Observational memory ${operationType} ${stage} failed: ${error}`)
    });
    this.abort();
  }
  async processStream(response, requestContextInput) {
    const state = this.createStreamState();
    const requestContext = await this.buildRequestContext(requestContextInput);
    this.currentOperationId += 1;
    this.emit({ type: "agent_start" });
    let result;
    let error = false;
    let aborted = false;
    for await (const chunk of response.fullStream) {
      result = await this.processStreamChunk(state, chunk, requestContext);
      if (chunk.type === "error") {
        error = true;
      }
      if (chunk.type === "abort") {
        aborted = true;
      }
      if (result || chunk.type === "finish" || chunk.type === "error" || chunk.type === "abort" || chunk.type === "tool-call-suspended" || this.abortRequested) {
        result ??= this.finishStreamState(state);
        break;
      }
    }
    result ??= this.finishStreamState(state);
    this.emit({
      type: "agent_end",
      reason: error ? "error" : result.suspended ? "suspended" : aborted || this.abortRequested ? "aborted" : "complete"
    });
    this.currentRunId = null;
    this.currentTraceId = null;
    this.abortController = null;
    this.abortRequested = false;
    await this.drainFollowUpQueue();
    return result;
  }
  async processStreamChunk(state, chunk, requestContext) {
    if ("runId" in chunk && chunk.runId) {
      this.currentRunId = chunk.runId;
    }
    switch (chunk.type) {
      case "text-start": {
        const textIndex = state.currentMessage.content.length;
        state.currentMessage.content.push({ type: "text", text: "" });
        state.textContentById.set(chunk.payload.id, { index: textIndex, text: "" });
        this.emit({ type: "message_start", message: { ...state.currentMessage } });
        break;
      }
      case "text-delta": {
        const textState = state.textContentById.get(chunk.payload.id);
        if (textState) {
          textState.text += chunk.payload.text;
          const textContent = state.currentMessage.content[textState.index];
          if (textContent && textContent.type === "text") {
            textContent.text = textState.text;
          }
          this.emit({ type: "message_update", message: { ...state.currentMessage } });
        }
        break;
      }
      case "reasoning-start": {
        const thinkingIndex = state.currentMessage.content.length;
        state.currentMessage.content.push({ type: "thinking", thinking: "" });
        state.thinkingContentById.set(chunk.payload.id, { index: thinkingIndex, text: "" });
        this.emit({ type: "message_update", message: { ...state.currentMessage } });
        break;
      }
      case "reasoning-delta": {
        const thinkingState = state.thinkingContentById.get(chunk.payload.id);
        if (thinkingState) {
          thinkingState.text += chunk.payload.text;
          const thinkingContent = state.currentMessage.content[thinkingState.index];
          if (thinkingContent && thinkingContent.type === "thinking") {
            thinkingContent.thinking = thinkingState.text;
          }
          this.emit({ type: "message_update", message: { ...state.currentMessage } });
        }
        break;
      }
      case "tool-call-input-streaming-start": {
        const { toolCallId, toolName } = chunk.payload;
        this.emit({ type: "tool_input_start", toolCallId, toolName });
        break;
      }
      case "tool-call-delta": {
        const { toolCallId, argsTextDelta, toolName } = chunk.payload;
        const transform = chunkVPZGHUNF_cjs.getTransformedToolPayload(chunk.metadata, "display", "input-delta");
        if (!transform?.suppress) {
          this.emit({
            type: "tool_input_delta",
            toolCallId,
            argsTextDelta: chunkVPZGHUNF_cjs.hasTransformedToolPayload(transform) ? transform.transformed : argsTextDelta,
            toolName
          });
        }
        break;
      }
      case "tool-call-input-streaming-end": {
        const { toolCallId } = chunk.payload;
        this.emit({ type: "tool_input_end", toolCallId });
        break;
      }
      case "tool-call": {
        const toolCall = chunk.payload;
        const args = getDisplayTransform(chunk.metadata, "input-available", toolCall.args);
        state.currentMessage.content.push({
          type: "tool_call",
          id: toolCall.toolCallId,
          name: toolCall.toolName,
          args
        });
        this.emit({
          type: "tool_start",
          toolCallId: toolCall.toolCallId,
          toolName: toolCall.toolName,
          args
        });
        this.emit({ type: "message_update", message: { ...state.currentMessage } });
        break;
      }
      case "tool-result": {
        const toolResult = chunk.payload;
        const providerMetadata = toolResult.providerMetadata;
        const result = getDisplayTransform(chunk.metadata, "output-available", toolResult.result);
        state.currentMessage.content.push({
          type: "tool_result",
          id: toolResult.toolCallId,
          name: toolResult.toolName,
          result,
          isError: toolResult.isError ?? false,
          ...providerMetadata ? { providerMetadata } : {}
        });
        this.emit({
          type: "tool_end",
          toolCallId: toolResult.toolCallId,
          result,
          isError: toolResult.isError ?? false,
          ...providerMetadata ? { providerMetadata } : {}
        });
        this.emit({ type: "message_update", message: { ...state.currentMessage } });
        break;
      }
      case "tool-error": {
        const toolError = chunk.payload;
        this.emit({
          type: "tool_end",
          toolCallId: toolError.toolCallId,
          result: getDisplayTransform(chunk.metadata, "error", toolError.error),
          isError: true
        });
        break;
      }
      case "tool-call-approval": {
        const toolCallId = chunk.payload.toolCallId;
        const toolName = chunk.payload.toolName;
        const approvalTransform = chunkVPZGHUNF_cjs.getTransformedToolPayload(chunk.metadata, "display", "approval");
        const toolArgs = chunkVPZGHUNF_cjs.hasTransformedToolPayload(approvalTransform) ? approvalTransform.transformed : getDisplayTransform(chunk.metadata, "input-available", chunk.payload.args);
        const policy = this.resolveToolApproval(toolName);
        if (policy === "allow") {
          await this.handleToolApprove({ toolCallId, requestContext });
          break;
        }
        if (policy === "deny") {
          await this.handleToolDecline({ toolCallId, requestContext });
          break;
        }
        this.pendingApprovalToolName = toolName;
        this.emit({ type: "tool_approval_required", toolCallId, toolName, args: toolArgs });
        const approval = await new Promise(
          (resolve) => {
            this.pendingApprovalResolve = resolve;
          }
        );
        this.pendingApprovalToolName = null;
        if (approval.decision === "approve") {
          await this.handleToolApprove({ toolCallId, requestContext: approval.requestContext ?? requestContext });
        } else {
          await this.handleToolDecline({ toolCallId, requestContext: approval.requestContext ?? requestContext });
        }
        break;
      }
      case "tool-call-suspended": {
        const suspToolCallId = chunk.payload.toolCallId;
        const suspToolName = chunk.payload.toolName;
        const suspArgs = getDisplayTransform(chunk.metadata, "input-available", chunk.payload.args);
        const suspPayload = getDisplayTransform(chunk.metadata, "suspend", chunk.payload.suspendPayload);
        const suspResumeSchema = chunk.payload.resumeSchema;
        this.emit({
          type: "tool_suspended",
          toolCallId: suspToolCallId,
          toolName: suspToolName,
          args: suspArgs,
          suspendPayload: suspPayload,
          resumeSchema: suspResumeSchema
        });
        this.pendingSuspensionRunId = this.currentRunId;
        this.pendingSuspensionToolCallId = suspToolCallId;
        state.isSuspended = true;
        break;
      }
      case "error": {
        const streamError = chunk.payload.error instanceof Error ? chunk.payload.error : new Error(String(chunk.payload.error));
        this.emit({ type: "error", error: streamError });
        break;
      }
      case "step-finish": {
        const usage = chunk.payload?.output?.usage;
        if (usage) {
          const usageRecord = usage;
          const promptTokens = getUsageNumber(usageRecord, "promptTokens") ?? getUsageNumber(usageRecord, "inputTokens") ?? 0;
          const completionTokens = getUsageNumber(usageRecord, "completionTokens") ?? getUsageNumber(usageRecord, "outputTokens") ?? 0;
          const totalTokens = getUsageNumber(usageRecord, "totalTokens") ?? promptTokens + completionTokens;
          const stepUsage = {
            promptTokens,
            completionTokens,
            totalTokens
          };
          addOptionalUsageField(stepUsage, "reasoningTokens", getUsageNumber(usageRecord, "reasoningTokens"));
          addOptionalUsageField(stepUsage, "cachedInputTokens", getUsageNumber(usageRecord, "cachedInputTokens"));
          addOptionalUsageField(
            stepUsage,
            "cacheCreationInputTokens",
            getUsageNumber(usageRecord, "cacheCreationInputTokens")
          );
          if (usageRecord.raw !== void 0) {
            stepUsage.raw = usageRecord.raw;
          }
          this.tokenUsage.promptTokens += promptTokens;
          this.tokenUsage.completionTokens += completionTokens;
          this.tokenUsage.totalTokens += totalTokens;
          addOptionalUsageField(this.tokenUsage, "reasoningTokens", stepUsage.reasoningTokens);
          addOptionalUsageField(this.tokenUsage, "cachedInputTokens", stepUsage.cachedInputTokens);
          addOptionalUsageField(this.tokenUsage, "cacheCreationInputTokens", stepUsage.cacheCreationInputTokens);
          if (stepUsage.raw !== void 0) {
            this.tokenUsage.raw = stepUsage.raw;
          }
          this.persistTokenUsage().catch(() => {
          });
          this.emit({ type: "usage_update", usage: stepUsage });
        }
        break;
      }
      case "finish": {
        const finishReason = chunk.payload.stepResult?.reason;
        if (finishReason === "stop" || finishReason === "end-turn") {
          state.currentMessage.stopReason = "complete";
        } else if (finishReason === "tool-calls") {
          state.currentMessage.stopReason = "tool_use";
        } else {
          state.currentMessage.stopReason = "complete";
        }
        break;
      }
      // Observational Memory data parts
      // NOTE: OM data parts arrive as { type, data: { ... } } — NOT { type, payload }
      case "data-om-status": {
        const d = chunk.data;
        if (d?.windows) {
          const w = d.windows;
          const active = w.active ?? {};
          const msgs = active.messages ?? {};
          const obs = active.observations ?? {};
          const buffObs = w.buffered?.observations ?? {};
          const buffRef = w.buffered?.reflection ?? {};
          this.emit({
            type: "om_status",
            windows: {
              active: {
                messages: { tokens: msgs.tokens ?? 0, threshold: msgs.threshold ?? 0 },
                observations: { tokens: obs.tokens ?? 0, threshold: obs.threshold ?? 0 }
              },
              buffered: {
                observations: {
                  status: buffObs.status ?? "idle",
                  chunks: buffObs.chunks ?? 0,
                  messageTokens: buffObs.messageTokens ?? 0,
                  projectedMessageRemoval: buffObs.projectedMessageRemoval ?? 0,
                  observationTokens: buffObs.observationTokens ?? 0
                },
                reflection: {
                  status: buffRef.status ?? "idle",
                  inputObservationTokens: buffRef.inputObservationTokens ?? 0,
                  observationTokens: buffRef.observationTokens ?? 0
                }
              }
            },
            recordId: d.recordId ?? "",
            threadId: d.threadId ?? "",
            stepNumber: d.stepNumber ?? 0,
            generationCount: d.generationCount ?? 0
          });
        }
        break;
      }
      case "data-om-observation-start": {
        const payload = chunk.data;
        if (payload && payload.cycleId) {
          if (payload.operationType === "observation") {
            this.emit({
              type: "om_observation_start",
              cycleId: payload.cycleId,
              operationType: payload.operationType,
              tokensToObserve: payload.tokensToObserve ?? 0
            });
          } else if (payload.operationType === "reflection") {
            this.emit({
              type: "om_reflection_start",
              cycleId: payload.cycleId,
              tokensToReflect: payload.tokensToObserve ?? 0
            });
          }
        }
        break;
      }
      case "data-om-observation-end": {
        const payload = chunk.data;
        if (payload && payload.cycleId) {
          if (payload.operationType === "reflection") {
            this.emit({
              type: "om_reflection_end",
              cycleId: payload.cycleId,
              durationMs: payload.durationMs ?? 0,
              compressedTokens: payload.observationTokens ?? 0,
              observations: payload.observations
            });
          } else {
            this.emit({
              type: "om_observation_end",
              cycleId: payload.cycleId,
              durationMs: payload.durationMs ?? 0,
              tokensObserved: payload.tokensObserved ?? 0,
              observationTokens: payload.observationTokens ?? 0,
              observations: payload.observations,
              currentTask: payload.currentTask,
              suggestedResponse: payload.suggestedResponse
            });
          }
        }
        break;
      }
      case "data-om-observation-failed": {
        const payload = chunk.data;
        if (payload) {
          const operationType = payload.operationType === "reflection" ? "reflection" : "observation";
          const error = payload.error ?? "Unknown error";
          if (operationType === "reflection") {
            this.emit({
              type: "om_reflection_failed",
              cycleId: payload.cycleId ?? "unknown",
              error,
              durationMs: payload.durationMs ?? 0
            });
          } else {
            this.emit({
              type: "om_observation_failed",
              cycleId: payload.cycleId ?? "unknown",
              error,
              durationMs: payload.durationMs ?? 0
            });
          }
          this.abortForOmFailure({ operationType, stage: "run", error });
          return { message: state.currentMessage };
        }
        break;
      }
      // Async buffering lifecycle
      case "data-om-buffering-start": {
        const payload = chunk.data;
        if (payload && payload.cycleId) {
          this.emit({
            type: "om_buffering_start",
            cycleId: payload.cycleId,
            operationType: payload.operationType ?? "observation",
            tokensToBuffer: payload.tokensToBuffer ?? 0
          });
        }
        break;
      }
      case "data-om-buffering-end": {
        const payload = chunk.data;
        if (payload && payload.cycleId) {
          this.emit({
            type: "om_buffering_end",
            cycleId: payload.cycleId,
            operationType: payload.operationType ?? "observation",
            tokensBuffered: payload.tokensBuffered ?? 0,
            bufferedTokens: payload.bufferedTokens ?? 0,
            observations: payload.observations
          });
        }
        break;
      }
      case "data-om-buffering-failed": {
        const payload = chunk.data;
        if (payload) {
          const operationType = payload.operationType ?? "observation";
          const error = payload.error ?? "Unknown error";
          this.emit({
            type: "om_buffering_failed",
            cycleId: payload.cycleId,
            operationType,
            error
          });
          this.abortForOmFailure({ operationType, stage: "buffering", error });
          return { message: state.currentMessage };
        }
        break;
      }
      case "data-user-message": {
        const payload = chunk.data;
        const message = payload ? toUserSignalMessage(payload) : void 0;
        if (message) {
          if (state.currentMessage.content.length > 0) {
            state.currentMessage.stopReason ??= "complete";
            this.emit({ type: "message_end", message: { ...state.currentMessage } });
            state.currentMessage = {
              id: this.generateId(),
              role: "assistant",
              content: [],
              createdAt: /* @__PURE__ */ new Date()
            };
            state.textContentById.clear();
            state.thinkingContentById.clear();
          }
          this.emit({ type: "message_start", message });
          this.emit({ type: "message_end", message });
        }
        break;
      }
      case "data-system-reminder": {
        const payload = chunk.data;
        const reminder = payload ? toSystemReminderContent(payload) : void 0;
        if (reminder) {
          state.currentMessage.content.push(reminder);
          this.emit({ type: "message_update", message: state.currentMessage });
        }
        break;
      }
      case "data-om-activation": {
        const payload = chunk.data;
        if (payload && payload.cycleId) {
          this.emit({
            type: "om_activation",
            cycleId: payload.cycleId,
            operationType: payload.operationType ?? "observation",
            chunksActivated: payload.chunksActivated ?? 0,
            tokensActivated: payload.tokensActivated ?? 0,
            observationTokens: payload.observationTokens ?? 0,
            messagesActivated: payload.messagesActivated ?? 0,
            generationCount: payload.generationCount ?? 0,
            triggeredBy: payload.triggeredBy,
            lastActivityAt: payload.lastActivityAt,
            ttlExpiredMs: payload.ttlExpiredMs,
            activateAfterIdle: typeof payload.config?.activateAfterIdle === "number" ? payload.config.activateAfterIdle : void 0,
            previousModel: payload.previousModel,
            currentModel: payload.currentModel
          });
        }
        break;
      }
      case "data-om-thread-update": {
        const payload = chunk.data;
        if (payload && payload.newTitle) {
          this.emit({
            type: "om_thread_title_updated",
            cycleId: payload.cycleId ?? "unknown",
            threadId: payload.threadId ?? this.currentThreadId ?? "unknown",
            oldTitle: payload.oldTitle,
            newTitle: payload.newTitle
          });
        }
        break;
      }
      // Sandbox streaming data chunks (from workspace execute_command tool)
      case "data-sandbox-stdout": {
        const d = chunk.data;
        if (d?.output && d?.toolCallId) {
          this.emit({ type: "shell_output", toolCallId: d.toolCallId, output: d.output, stream: "stdout" });
        }
        break;
      }
      case "data-sandbox-stderr": {
        const d = chunk.data;
        if (d?.output && d?.toolCallId) {
          this.emit({ type: "shell_output", toolCallId: d.toolCallId, output: d.output, stream: "stderr" });
        }
        break;
      }
    }
  }
  finishStreamState(state) {
    this.emit({ type: "message_end", message: state.currentMessage });
    return { message: state.currentMessage, suspended: state.isSuspended || void 0 };
  }
  // ===========================================================================
  // Control
  // ===========================================================================
  /**
   * Abort the current operation.
   */
  abort() {
    this.abortRequested = true;
    try {
      this.agentThreadSubscription?.abort();
    } catch {
    }
    if (this.abortController) {
      try {
        this.abortController.abort();
      } catch {
      }
      this.abortController = null;
    }
  }
  /**
   * Steer the agent mid-stream: aborts current run and sends a new message.
   */
  async steer({ content, requestContext }) {
    this.abort();
    this.followUpQueue = [];
    this.emit({ type: "follow_up_queued", count: 0 });
    await this.sendMessage({ content, requestContext });
  }
  /**
   * Queue a follow-up message to be processed after the current operation completes.
   */
  async followUp({ content, requestContext }) {
    if (this.isRunning()) {
      this.followUpQueue.push({ content, requestContext });
      this.emit({ type: "follow_up_queued", count: this.followUpQueue.length });
    } else {
      await this.sendMessage({ content, requestContext });
    }
  }
  getFollowUpCount() {
    return this.followUpQueue.length;
  }
  isRunning() {
    return this.abortController !== null;
  }
  getCurrentRunId() {
    return this.agentThreadSubscription?.activeRunId() ?? this.currentRunId;
  }
  isCurrentThreadStreamActive() {
    return this.agentThreadSubscription?.activeRunId() !== null && this.agentThreadSubscription?.activeRunId() !== void 0;
  }
  /**
   * Resolve once the current thread's stream is fully idle.
   *
   * After `abort()` is called the run's status can still be `'running'` for a
   * few microtasks while the underlying model stream finalizes. Callers that
   * need to send a fresh signal after an abort (e.g. plan approval → mode
   * switch → trigger reminder) should await this before calling `sendSignal`
   * to avoid the new signal being queued onto the dying run, which would then
   * be drained with the previous run's already-aborted abortSignal.
   */
  async waitForCurrentThreadStreamIdle() {
    while (this.isCurrentThreadStreamActive() || this.currentRunId !== null) {
      await new Promise((resolve) => setTimeout(resolve, 0));
    }
  }
  getCurrentTraceId() {
    return this.currentTraceId;
  }
  getSubagentDisplayName(agentType) {
    return this.config.subagents?.find((subagent) => subagent.id === agentType)?.name;
  }
  // ===========================================================================
  // Display State
  // ===========================================================================
  /**
   * Returns a read-only snapshot of the canonical display state.
   * UIs should use this to render instead of building up state from raw events.
   */
  getDisplayState() {
    return this.displayState;
  }
  /**
   * Restore task display state after a UI replays persisted task tool history.
   * This updates the Harness-owned display snapshot without emitting a live
   * `task_updated` event, since no task tool just ran.
   */
  restoreDisplayTasks(tasks) {
    this.displayState.previousTasks = [...this.displayState.tasks];
    this.displayState.tasks = [...tasks];
    this.dispatchDisplayStateChanged(false);
  }
  /**
   * Reset display state fields that are scoped to a thread.
   * Called on thread switch/creation.
   */
  resetThreadDisplayState() {
    this.displayState.activeTools = /* @__PURE__ */ new Map();
    this.displayState.toolInputBuffers = /* @__PURE__ */ new Map();
    this.displayState.pendingApproval = null;
    this.displayState.pendingSuspension = null;
    this.displayState.pendingQuestion = null;
    this.displayState.pendingPlanApproval = null;
    this.displayState.activeSubagents = /* @__PURE__ */ new Map();
    this.displayState.currentMessage = null;
    this.followUpQueue = [];
    this.displayState.queuedFollowUps = 0;
    this.displayState.modifiedFiles = /* @__PURE__ */ new Map();
    this.displayState.tasks = [];
    this.displayState.previousTasks = [];
    this.displayState.omProgress = defaultOMProgressState();
    this.displayState.bufferingMessages = false;
    this.displayState.bufferingObservations = false;
  }
  /**
   * Respond to a pending tool approval from the UI.
   * "always_allow_category" grants the tool's category for the rest of the session, then approves.
   */
  respondToToolApproval({
    decision,
    requestContext
  }) {
    if (!this.pendingApprovalResolve) return;
    if (decision === "always_allow_category") {
      const tn = this.pendingApprovalToolName;
      if (tn) {
        const category = this.getToolCategory({ toolName: tn });
        if (category) {
          this.grantSessionCategory({ category });
        }
      }
      this.pendingApprovalResolve({ decision: "approve", requestContext });
    } else {
      this.pendingApprovalResolve({ decision, requestContext });
    }
    this.pendingApprovalResolve = null;
  }
  /**
   * Respond to a pending tool suspension from the UI.
   * Provides resume data so the suspended tool can continue execution.
   */
  async respondToToolSuspension({
    resumeData,
    requestContext
  }) {
    if (!this.pendingSuspensionRunId) return;
    try {
      await this.handleToolResume({
        resumeData,
        requestContext
      });
    } catch (error) {
      const err = error instanceof Error ? error : new Error(String(error));
      this.emit({ type: "error", error: err });
      this.emit({ type: "agent_end", reason: "error" });
    }
  }
  // ===========================================================================
  // Question & Plan Approval
  // ===========================================================================
  /**
   * Register a pending question resolver.
   * Called by agent tools (e.g., ask_user) to pause execution until the UI responds.
   */
  registerQuestion({
    questionId,
    resolve
  }) {
    this.pendingQuestions.set(questionId, resolve);
  }
  /**
   * Resolve a pending question with the user's answer.
   * Called by the UI when the user responds to a question dialog.
   */
  respondToQuestion({ questionId, answer }) {
    const resolve = this.pendingQuestions.get(questionId);
    if (resolve) {
      this.pendingQuestions.delete(questionId);
      resolve(answer);
    }
  }
  /**
   * Register a pending plan approval resolver.
   * Called by agent tools (e.g., submit_plan) to pause execution until approval.
   */
  registerPlanApproval({
    planId,
    resolve
  }) {
    this.pendingPlanApprovals.set(planId, resolve);
  }
  /**
   * Respond to a pending plan approval.
   * On approval: resolves the suspended plan tool, then switches to the default mode.
   * On rejection: resolves with feedback (stays in current mode).
   */
  async respondToPlanApproval({
    planId,
    response
  }) {
    const resolve = this.pendingPlanApprovals.get(planId);
    if (!resolve) return;
    this.pendingPlanApprovals.delete(planId);
    resolve(response);
    if (response.action === "approved") {
      const defaultMode = this.config.modes.find((m) => m.default) ?? this.config.modes[0];
      if (defaultMode && defaultMode.id !== this.currentModeId) {
        await new Promise((resolveTimeout) => setTimeout(resolveTimeout, 0));
        await this.switchMode({ modeId: defaultMode.id });
        await this.waitForCurrentThreadStreamIdle();
      }
    }
  }
  async handleToolApprove({
    toolCallId,
    requestContext: requestContextInput
  }) {
    if (!this.currentRunId) {
      throw new Error("No active run to approve tool call for");
    }
    const agent = this.getCurrentAgent();
    if (!this.abortController) {
      this.abortController = new AbortController();
    }
    const requestContext = await this.buildRequestContext(requestContextInput);
    const isYolo = this.state.yolo === true;
    await agent.approveToolCall({
      runId: this.currentRunId,
      toolCallId,
      requireToolApproval: !isYolo,
      memory: this.currentThreadId ? { thread: this.currentThreadId, resource: this.resourceId } : void 0,
      abortSignal: this.abortController.signal,
      requestContext,
      toolsets: await this.buildToolsets(requestContext)
    });
  }
  async handleToolDecline({
    toolCallId,
    requestContext: requestContextInput
  }) {
    if (!this.currentRunId) {
      throw new Error("No active run to decline tool call for");
    }
    const agent = this.getCurrentAgent();
    if (!this.abortController) {
      this.abortController = new AbortController();
    }
    const requestContext = await this.buildRequestContext(requestContextInput);
    const isYolo = this.state.yolo === true;
    await agent.declineToolCall({
      runId: this.currentRunId,
      toolCallId,
      requireToolApproval: !isYolo,
      memory: this.currentThreadId ? { thread: this.currentThreadId, resource: this.resourceId } : void 0,
      abortSignal: this.abortController.signal,
      requestContext,
      toolsets: await this.buildToolsets(requestContext)
    });
  }
  async handleToolResume({
    resumeData,
    requestContext: requestContextInput
  }) {
    if (!this.pendingSuspensionRunId) {
      throw new Error("No active suspension to resume");
    }
    const agent = this.getCurrentAgent();
    if (!this.abortController) {
      this.abortController = new AbortController();
    }
    const requestContext = await this.buildRequestContext(requestContextInput);
    const isYolo = this.state.yolo === true;
    const output = await agent.resumeStream(resumeData, {
      runId: this.pendingSuspensionRunId,
      toolCallId: this.pendingSuspensionToolCallId ?? void 0,
      requireToolApproval: !isYolo,
      memory: this.currentThreadId ? { thread: this.currentThreadId, resource: this.resourceId } : void 0,
      abortSignal: this.abortController.signal,
      requestContext,
      toolsets: await this.buildToolsets(requestContext)
    });
    await this.processStream(output, requestContext);
    this.pendingSuspensionRunId = null;
    this.pendingSuspensionToolCallId = null;
  }
  // ===========================================================================
  // Event System
  // ===========================================================================
  /**
   * Subscribe to harness events. Returns an unsubscribe function.
   */
  subscribe(listener) {
    this.listeners.push(listener);
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index !== -1) {
        this.listeners.splice(index, 1);
      }
    };
  }
  /**
   * Subscribe to coalesced display state snapshots.
   *
   * Use this for UI rendering paths that only need the latest display state.
   * Raw event consumers should continue to use `subscribe()`.
   */
  subscribeDisplayState(listener, options = {}) {
    const scheduler = new DisplayStateScheduler(
      listener,
      options.windowMs ?? DEFAULT_DISPLAY_STATE_SUBSCRIPTION_OPTIONS.windowMs,
      options.maxWaitMs ?? DEFAULT_DISPLAY_STATE_SUBSCRIPTION_OPTIONS.maxWaitMs
    );
    this.displayStateSchedulers.add(scheduler);
    return () => {
      this.displayStateSchedulers.delete(scheduler);
      scheduler.dispose();
    };
  }
  emit(event) {
    this.applyDisplayStateUpdate(event);
    this.dispatchToListeners(event);
    if (event.type !== "display_state_changed") {
      const isCritical = CRITICAL_DISPLAY_STATE_EVENT_TYPES.has(event.type);
      this.dispatchDisplayStateChanged(isCritical);
    }
  }
  dispatchDisplayStateChanged(isCritical) {
    this.dispatchToListeners({
      type: "display_state_changed",
      displayState: this.displayState
    });
    if (this.displayStateSchedulers.size > 0) {
      for (const scheduler of Array.from(this.displayStateSchedulers)) {
        scheduler.notify(this.displayState, isCritical);
      }
    }
  }
  dispatchToListeners(event) {
    for (const listener of this.listeners) {
      try {
        const result = listener(event);
        if (result && typeof result === "object" && "catch" in result) {
          result.catch((err) => console.error("Error in harness event listener:", err));
        }
      } catch (err) {
        console.error("Error in harness event listener:", err);
      }
    }
  }
  /**
   * Apply a display state update based on an incoming event.
   * This is the centralized state machine that keeps HarnessDisplayState in sync
   * with every event the Harness emits.
   */
  applyDisplayStateUpdate(event) {
    const ds = this.displayState;
    switch (event.type) {
      // ── Agent lifecycle ────────────────────────────────────────────────
      case "agent_start":
        ds.isRunning = true;
        ds.activeTools = /* @__PURE__ */ new Map();
        ds.toolInputBuffers = /* @__PURE__ */ new Map();
        ds.currentMessage = null;
        ds.pendingApproval = null;
        ds.pendingSuspension = null;
        break;
      case "agent_end":
        ds.isRunning = false;
        ds.pendingApproval = null;
        if (event.reason !== "suspended") {
          ds.pendingSuspension = null;
        }
        ds.pendingQuestion = null;
        ds.pendingPlanApproval = null;
        for (const [, tool] of ds.activeTools) {
          if (tool.status === "running" || tool.status === "streaming_input") {
            tool.status = "error";
          }
        }
        ds.activeSubagents = /* @__PURE__ */ new Map();
        break;
      // ── Message streaming ──────────────────────────────────────────────
      case "message_start":
        ds.currentMessage = event.message;
        break;
      case "message_update":
        ds.currentMessage = event.message;
        break;
      case "message_end":
        ds.currentMessage = event.message;
        break;
      // ── Tool lifecycle ─────────────────────────────────────────────────
      case "tool_input_start": {
        ds.toolInputBuffers.set(event.toolCallId, { text: "", toolName: event.toolName });
        const existing = ds.activeTools.get(event.toolCallId);
        if (existing) {
          existing.status = "streaming_input";
        } else {
          ds.activeTools.set(event.toolCallId, {
            name: event.toolName,
            args: {},
            status: "streaming_input"
          });
        }
        break;
      }
      case "tool_input_delta": {
        const buf = ds.toolInputBuffers.get(event.toolCallId);
        if (buf) {
          buf.text += event.argsTextDelta;
        }
        break;
      }
      case "tool_input_end":
        ds.toolInputBuffers.delete(event.toolCallId);
        break;
      case "tool_start": {
        const existingTool = ds.activeTools.get(event.toolCallId);
        if (existingTool) {
          existingTool.name = event.toolName;
          existingTool.args = event.args;
          existingTool.status = "running";
        } else {
          ds.activeTools.set(event.toolCallId, {
            name: event.toolName,
            args: event.args,
            status: "running"
          });
        }
        break;
      }
      case "tool_update": {
        const tool = ds.activeTools.get(event.toolCallId);
        if (tool) {
          tool.partialResult = typeof event.partialResult === "string" ? event.partialResult : chunkNLACGSDP_cjs.safeStringify(event.partialResult);
        }
        break;
      }
      case "tool_end": {
        const endedTool = ds.activeTools.get(event.toolCallId);
        if (endedTool) {
          endedTool.status = event.isError ? "error" : "completed";
          endedTool.result = event.result;
          endedTool.isError = event.isError;
        }
        if (!event.isError) {
          const FILE_TOOLS = ["string_replace_lsp", "write_file", "ast_smart_edit"];
          const toolState = ds.activeTools.get(event.toolCallId);
          if (toolState && FILE_TOOLS.includes(toolState.name)) {
            const toolArgs = toolState.args;
            const filePath = toolArgs?.path;
            if (filePath) {
              const existing = ds.modifiedFiles.get(filePath);
              if (existing) {
                existing.operations.push(toolState.name);
              } else {
                ds.modifiedFiles.set(filePath, {
                  operations: [toolState.name],
                  firstModified: /* @__PURE__ */ new Date()
                });
              }
            }
          }
        }
        break;
      }
      case "shell_output": {
        const shellTool = ds.activeTools.get(event.toolCallId);
        if (shellTool) {
          shellTool.shellOutput = (shellTool.shellOutput ?? "") + event.output;
        }
        break;
      }
      case "tool_approval_required":
        ds.pendingApproval = {
          toolCallId: event.toolCallId,
          toolName: event.toolName,
          args: event.args
        };
        break;
      case "tool_suspended":
        ds.pendingSuspension = {
          toolCallId: event.toolCallId,
          toolName: event.toolName,
          args: event.args,
          suspendPayload: event.suspendPayload,
          resumeSchema: event.resumeSchema
        };
        break;
      // ── Interactive prompts ────────────────────────────────────────────
      case "ask_question":
        ds.pendingQuestion = {
          questionId: event.questionId,
          question: event.question,
          options: event.options,
          selectionMode: event.selectionMode
        };
        break;
      case "plan_approval_required":
        ds.pendingPlanApproval = {
          planId: event.planId,
          title: event.title,
          plan: event.plan
        };
        break;
      case "plan_approved":
        ds.pendingPlanApproval = null;
        break;
      // ── Subagent tracking ──────────────────────────────────────────────
      case "subagent_start": {
        const displayName = this.getSubagentDisplayName(event.agentType);
        ds.activeSubagents.set(event.toolCallId, {
          agentType: event.agentType,
          ...displayName !== void 0 ? { displayName } : {},
          task: event.task,
          modelId: event.modelId,
          forked: event.forked,
          toolCalls: [],
          textDelta: "",
          status: "running"
        });
        break;
      }
      case "subagent_text_delta": {
        const sub = ds.activeSubagents.get(event.toolCallId);
        if (sub) {
          sub.textDelta += event.textDelta;
        }
        break;
      }
      case "subagent_tool_start": {
        const subAgent = ds.activeSubagents.get(event.toolCallId);
        if (subAgent) {
          subAgent.toolCalls.push({ name: event.subToolName, isError: false });
        }
        break;
      }
      case "subagent_tool_end": {
        const subTool = ds.activeSubagents.get(event.toolCallId);
        if (subTool) {
          const tc = subTool.toolCalls.find((t) => t.name === event.subToolName && !t.isError);
          if (tc) {
            tc.isError = event.isError;
          }
        }
        break;
      }
      case "subagent_end": {
        const endedSub = ds.activeSubagents.get(event.toolCallId);
        if (endedSub) {
          endedSub.status = event.isError ? "error" : "completed";
          endedSub.durationMs = event.durationMs;
          endedSub.result = event.result;
        }
        break;
      }
      // ── Observational Memory ───────────────────────────────────────────
      case "om_status": {
        const w = event.windows;
        ds.omProgress.pendingTokens = w.active.messages.tokens;
        ds.omProgress.threshold = w.active.messages.threshold;
        ds.omProgress.thresholdPercent = w.active.messages.threshold > 0 ? w.active.messages.tokens / w.active.messages.threshold * 100 : 0;
        ds.omProgress.observationTokens = w.active.observations.tokens;
        ds.omProgress.reflectionThreshold = w.active.observations.threshold;
        ds.omProgress.reflectionThresholdPercent = w.active.observations.threshold > 0 ? w.active.observations.tokens / w.active.observations.threshold * 100 : 0;
        ds.omProgress.buffered = {
          observations: { ...w.buffered.observations },
          reflection: { ...w.buffered.reflection }
        };
        ds.omProgress.generationCount = event.generationCount;
        ds.omProgress.stepNumber = event.stepNumber;
        ds.bufferingMessages = w.buffered.observations.status === "running";
        ds.bufferingObservations = w.buffered.reflection.status === "running";
        break;
      }
      case "om_observation_start":
        ds.omProgress.status = "observing";
        ds.omProgress.cycleId = event.cycleId;
        ds.omProgress.startTime = Date.now();
        break;
      case "om_observation_end":
        ds.omProgress.status = "idle";
        ds.omProgress.cycleId = void 0;
        ds.omProgress.startTime = void 0;
        ds.omProgress.observationTokens = event.observationTokens;
        ds.omProgress.pendingTokens = 0;
        ds.omProgress.thresholdPercent = 0;
        break;
      case "om_observation_failed":
        ds.omProgress.status = "idle";
        ds.omProgress.cycleId = void 0;
        ds.omProgress.startTime = void 0;
        break;
      case "om_reflection_start":
        ds.omProgress.status = "reflecting";
        ds.omProgress.cycleId = event.cycleId;
        ds.omProgress.startTime = Date.now();
        ds.omProgress.preReflectionTokens = ds.omProgress.observationTokens;
        ds.omProgress.observationTokens = event.tokensToReflect;
        ds.omProgress.reflectionThresholdPercent = ds.omProgress.reflectionThreshold > 0 ? event.tokensToReflect / ds.omProgress.reflectionThreshold * 100 : 0;
        break;
      case "om_reflection_end":
        ds.omProgress.status = "idle";
        ds.omProgress.cycleId = void 0;
        ds.omProgress.startTime = void 0;
        ds.omProgress.observationTokens = event.compressedTokens;
        ds.omProgress.reflectionThresholdPercent = ds.omProgress.reflectionThreshold > 0 ? event.compressedTokens / ds.omProgress.reflectionThreshold * 100 : 0;
        break;
      case "om_reflection_failed":
        ds.omProgress.status = "idle";
        ds.omProgress.cycleId = void 0;
        ds.omProgress.startTime = void 0;
        break;
      case "om_buffering_start":
        if (event.operationType === "observation") {
          ds.bufferingMessages = true;
        } else {
          ds.bufferingObservations = true;
        }
        break;
      case "om_buffering_end":
        if (event.operationType === "observation") {
          ds.bufferingMessages = false;
        } else {
          ds.bufferingObservations = false;
        }
        break;
      case "om_buffering_failed":
        if (event.operationType === "observation") {
          ds.bufferingMessages = false;
        } else {
          ds.bufferingObservations = false;
        }
        break;
      case "om_activation":
        if (event.operationType === "observation") {
          ds.bufferingMessages = false;
        } else {
          ds.bufferingObservations = false;
        }
        break;
      // ── Token usage ────────────────────────────────────────────────────
      case "usage_update":
        ds.tokenUsage = { ...this.tokenUsage };
        break;
      // ── Tasks ──────────────────────────────────────────────────────────
      case "task_updated":
        ds.previousTasks = [...ds.tasks];
        ds.tasks = event.tasks;
        break;
      // ── Follow-up queue ────────────────────────────────────────────────
      case "follow_up_queued":
        ds.queuedFollowUps = event.count;
        break;
      // ── Thread lifecycle ───────────────────────────────────────────────
      case "thread_changed":
        this.resetThreadDisplayState();
        ds.tokenUsage = { ...this.tokenUsage };
        break;
      case "thread_created":
        this.resetThreadDisplayState();
        ds.tokenUsage = createEmptyTokenUsage();
        break;
      case "thread_deleted":
        if (!this.currentThreadId) {
          this.resetThreadDisplayState();
          ds.tokenUsage = createEmptyTokenUsage();
        }
        break;
      // ── State changes (for OM threshold overrides) ──────────────────────
      case "state_changed": {
        const keys = event.changedKeys;
        if (keys.includes("observationThreshold")) {
          const value = event.state.observationThreshold;
          if (typeof value === "number") {
            ds.omProgress.threshold = value;
            ds.omProgress.thresholdPercent = value > 0 ? ds.omProgress.pendingTokens / value * 100 : 0;
          }
        }
        if (keys.includes("reflectionThreshold")) {
          const value = event.state.reflectionThreshold;
          if (typeof value === "number") {
            ds.omProgress.reflectionThreshold = value;
            ds.omProgress.reflectionThresholdPercent = value > 0 ? ds.omProgress.observationTokens / value * 100 : 0;
          }
        }
        break;
      }
    }
  }
  // ===========================================================================
  // Runtime Context
  // ===========================================================================
  /**
   * Build the toolsets object that includes built-in harness tools (ask_user, submit_plan,
   * and optionally subagent) plus any user-configured tools.
   * Used by sendMessage, handleToolApprove, and handleToolDecline.
   */
  async buildToolsets(requestContext) {
    const builtInTools = {
      ask_user: askUserTool,
      submit_plan: submitPlanTool,
      task_write: taskWriteTool,
      task_update: taskUpdateTool,
      task_complete: taskCompleteTool,
      task_check: taskCheckTool
    };
    let resolvedHarnessTools = void 0;
    if (this.config.tools) {
      const tools = typeof this.config.tools === "function" ? await this.config.tools({ requestContext }) : this.config.tools;
      if (tools) {
        resolvedHarnessTools = { ...tools };
      }
    }
    if (this.config.subagents?.length && this.config.resolveModel) {
      const currentMode = this.getCurrentMode();
      const hasMemory = Boolean(this.config.memory);
      builtInTools.subagent = createSubagentTool({
        subagents: this.config.subagents,
        resolveModel: this.config.resolveModel,
        harnessTools: resolvedHarnessTools,
        fallbackModelId: currentMode?.defaultModelId,
        getParentModelId: () => this.getCurrentModelId(),
        // Resolved lazily so forked subagents see the current mode's agent
        // even if the mode switches between tool-call scheduling and execution.
        getParentAgent: () => {
          try {
            return this.getCurrentAgent();
          } catch {
            return void 0;
          }
        },
        // Only wired up when memory is configured. Clones at the memory layer
        // (not via Harness.cloneThread) so the parent thread stays the active
        // thread while the forked subagent runs on the clone.
        //
        // The clone is tagged with `forkedSubagent: true` + `parentThreadId` so
        // that thread pickers / startup flows can hide transient fork threads —
        // see `listThreads` (filtered by default).
        cloneThreadForFork: hasMemory ? async ({ sourceThreadId, resourceId, title }) => {
          const memory = await this.resolveMemory();
          const result = await memory.cloneThread({
            sourceThreadId,
            resourceId: resourceId ?? this.resourceId,
            title,
            metadata: {
              forkedSubagent: true,
              parentThreadId: sourceThreadId
            }
          });
          return { id: result.thread.id, resourceId: result.thread.resourceId };
        } : void 0,
        // Forks inherit the parent's toolsets verbatim so harness-injected
        // tools (`ask_user`, `submit_plan`, user-configured harness tools, etc.)
        // remain available inside the fork. The `subagent` entry itself is
        // deliberately kept — its schema/description are part of the parent's
        // prompt-cache prefix, and stripping it would invalidate the cache.
        // Recursive forking is blocked at runtime instead: see the patched
        // `subagent` execute that the forked tool path installs in `tools.ts`.
        getParentToolsets: (forkRequestContext) => this.buildToolsets(forkRequestContext ?? requestContext)
      });
    }
    if (this.config.disableBuiltinTools?.length) {
      for (const toolId of this.config.disableBuiltinTools) {
        delete builtInTools[toolId];
      }
    }
    const permissionRules = this.getPermissionRules();
    for (const [toolId, policy] of Object.entries(permissionRules.tools)) {
      if (policy === "deny") {
        delete builtInTools[toolId];
        delete resolvedHarnessTools?.[toolId];
      }
    }
    if (resolvedHarnessTools) {
      return { harnessBuiltIn: builtInTools, harness: resolvedHarnessTools };
    }
    return { harnessBuiltIn: builtInTools };
  }
  /**
   * Build request context for agent execution.
   * Tools can access harness state via requestContext.get('harness').
   */
  async buildRequestContext(requestContext) {
    requestContext ??= new chunk64VGLWFD_cjs.RequestContext();
    const harnessContext = {
      harnessId: this.id,
      state: this.getState(),
      getState: () => this.getState(),
      setState: (updates) => this.setState(updates),
      updateState: (updater) => this.updateState(updater),
      threadId: this.currentThreadId,
      resourceId: this.resourceId,
      modeId: this.currentModeId,
      abortSignal: this.abortController?.signal,
      workspace: this.workspace,
      emitEvent: (event) => this.emit(event),
      registerQuestion: (params) => this.registerQuestion(params),
      registerPlanApproval: (params) => this.registerPlanApproval(params),
      getSubagentModelId: (params) => this.getSubagentModelId(params)
    };
    requestContext.set("harness", harnessContext);
    if (this.workspaceFn) {
      const resolved = await Promise.resolve(this.workspaceFn({ requestContext }));
      harnessContext.workspace = resolved;
      this.workspace = resolved;
    }
    return requestContext;
  }
  /**
   * Resolve memory from config — handles both static instances and dynamic factory functions.
   */
  async resolveMemory() {
    const mem = this.config.memory;
    if (!mem) {
      throw new Error("Memory is not configured on this Harness");
    }
    if (typeof mem !== "function") {
      return mem;
    }
    const requestContext = await this.buildRequestContext();
    const resolved = await Promise.resolve(mem({ requestContext }));
    if (!resolved) {
      throw new Error("Dynamic memory factory returned empty value");
    }
    return resolved;
  }
  // ===========================================================================
  // Token Usage
  // ===========================================================================
  getTokenUsage() {
    return { ...this.tokenUsage };
  }
  async persistTokenUsage() {
    if (!this.currentThreadId || !this.config.storage) return;
    try {
      const memoryStorage = await this.getMemoryStorage();
      const thread = await memoryStorage.getThreadById({ threadId: this.currentThreadId });
      if (thread) {
        await memoryStorage.saveThread({
          thread: {
            ...thread,
            metadata: { ...thread.metadata, tokenUsage: this.tokenUsage },
            updatedAt: /* @__PURE__ */ new Date()
          }
        });
      }
    } catch {
    }
  }
  // ===========================================================================
  // Workspace
  // ===========================================================================
  getWorkspace() {
    return this.workspace;
  }
  /**
   * Eagerly resolve the workspace. For dynamic workspaces (factory function),
   * this triggers resolution and caches the result so getWorkspace() returns it.
   * Useful for code paths outside the request flow (e.g. slash commands).
   */
  async resolveWorkspace({
    requestContext
  } = {}) {
    if (this.workspace) return this.workspace;
    if (this.workspaceFn) {
      await this.buildRequestContext(requestContext);
      return this.workspace;
    }
    return void 0;
  }
  hasWorkspace() {
    return this.config.workspace !== void 0;
  }
  isWorkspaceReady() {
    if (this.workspaceFn) return true;
    return this.workspaceInitialized && this.workspace !== void 0;
  }
  async destroyWorkspace() {
    if (this.workspaceFn) return;
    if (this.workspace && this.workspaceInitialized) {
      try {
        this.emit({ type: "workspace_status_changed", status: "destroying" });
        await this.workspace.destroy();
        this.emit({ type: "workspace_status_changed", status: "destroyed" });
      } catch (error) {
        console.warn("Workspace destroy failed:", error);
      } finally {
        this.workspaceInitialized = false;
      }
    }
  }
  // ===========================================================================
  // Heartbeat Handlers
  // ===========================================================================
  startHeartbeats() {
    const handlers = this.config.heartbeatHandlers;
    if (!handlers?.length) return;
    for (const hb of handlers) {
      if (this.heartbeatTimers.has(hb.id)) continue;
      const run = async () => {
        try {
          await hb.handler();
        } catch (error) {
          console.error(`[Heartbeat:${hb.id}] failed:`, error);
        }
      };
      if (hb.immediate !== false) {
        void run();
      }
      const timer = setInterval(run, hb.intervalMs);
      timer.unref();
      this.heartbeatTimers.set(hb.id, { timer, shutdown: hb.shutdown });
    }
  }
  registerHeartbeat(handler) {
    void this.removeHeartbeat({ id: handler.id });
    const run = async () => {
      try {
        await handler.handler();
      } catch (error) {
        console.error(`[Heartbeat:${handler.id}] failed:`, error);
      }
    };
    if (handler.immediate !== false) {
      void run();
    }
    const timer = setInterval(run, handler.intervalMs);
    timer.unref();
    this.heartbeatTimers.set(handler.id, { timer, shutdown: handler.shutdown });
  }
  async removeHeartbeat({ id }) {
    const entry = this.heartbeatTimers.get(id);
    if (entry) {
      clearInterval(entry.timer);
      this.heartbeatTimers.delete(id);
      try {
        await entry.shutdown?.();
      } catch (error) {
        console.error(`[Heartbeat:${id}] shutdown failed:`, error);
      }
    }
  }
  async stopHeartbeats() {
    const entries = [...this.heartbeatTimers.entries()];
    this.heartbeatTimers.clear();
    for (const [id, entry] of entries) {
      clearInterval(entry.timer);
      try {
        await entry.shutdown?.();
      } catch (error) {
        console.error(`[Heartbeat:${id}] shutdown failed:`, error);
      }
    }
  }
  // ===========================================================================
  // Lifecycle
  // ===========================================================================
  async destroy() {
    this.cleanupAgentThreadSubscription();
    for (const scheduler of this.displayStateSchedulers) {
      scheduler.dispose();
    }
    this.displayStateSchedulers.clear();
    await this.stopHeartbeats();
    await this.destroyWorkspace();
  }
  // ===========================================================================
  // Session
  // ===========================================================================
  async getSession() {
    return {
      currentThreadId: this.currentThreadId,
      currentModeId: this.currentModeId,
      threads: await this.listThreads()
    };
  }
  // ===========================================================================
  // Utilities
  // ===========================================================================
  generateId() {
    if (this.config.idGenerator) {
      return this.config.idGenerator();
    }
    return `${Date.now()}-${Math.random().toString(36).slice(2, 11)}`;
  }
};

exports.Harness = Harness;
exports.askUserTool = askUserTool;
exports.assignTaskIds = assignTaskIds;
exports.defaultDisplayState = defaultDisplayState;
exports.defaultOMProgressState = defaultOMProgressState;
exports.parseSubagentMeta = parseSubagentMeta;
exports.submitPlanTool = submitPlanTool;
exports.taskCheckTool = taskCheckTool;
exports.taskCompleteTool = taskCompleteTool;
exports.taskUpdateTool = taskUpdateTool;
exports.taskWriteTool = taskWriteTool;
