// @mastra/core/dist/llm/index.cjs
'use strict';

var chunkDYWGJ6ZD_cjs = require('../chunk-DYWGJ6ZD.cjs');
var chunkXKOQ7YF6_cjs = require('../chunk-XKOQ7YF6.cjs');
var chunkCJZO62WO_cjs = require('../chunk-CJZO62WO.cjs');
var chunk7WVQHAW5_cjs = require('../chunk-7WVQHAW5.cjs');
var chunkJBYS3POB_cjs = require('../chunk-JBYS3POB.cjs');



Object.defineProperty(exports, "AzureOpenAIGateway", {
  enumerable: true,
  get: function () { return chunkDYWGJ6ZD_cjs.AzureOpenAIGateway; }
});
Object.defineProperty(exports, "EMBEDDING_MODELS", {
  enumerable: true,
  get: function () { return chunkDYWGJ6ZD_cjs.EMBEDDING_MODELS; }
});
Object.defineProperty(exports, "ModelRouterEmbeddingModel", {
  enumerable: true,
  get: function () { return chunkDYWGJ6ZD_cjs.ModelRouterEmbeddingModel; }
});
Object.defineProperty(exports, "ModelRouterLanguageModel", {
  enumerable: true,
  get: function () { return chunkDYWGJ6ZD_cjs.ModelRouterLanguageModel; }
});
Object.defineProperty(exports, "defaultGateways", {
  enumerable: true,
  get: function () { return chunkDYWGJ6ZD_cjs.defaultGateways; }
});
Object.defineProperty(exports, "resolveModelConfig", {
  enumerable: true,
  get: function () { return chunkDYWGJ6ZD_cjs.resolveModelConfig; }
});
Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "MastraGateway", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.MastraGateway; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "modelSupportsAttachments", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.modelSupportsAttachments; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkXKOQ7YF6_cjs.parseModelString; }
});
Object.defineProperty(exports, "NetlifyGateway", {
  enumerable: true,
  get: function () { return chunkCJZO62WO_cjs.NetlifyGateway; }
});
Object.defineProperty(exports, "ModelsDevGateway", {
  enumerable: true,
  get: function () { return chunk7WVQHAW5_cjs.ModelsDevGateway; }
});
Object.defineProperty(exports, "GATEWAY_AUTH_HEADER", {
  enumerable: true,
  get: function () { return chunkJBYS3POB_cjs.GATEWAY_AUTH_HEADER; }
});
Object.defineProperty(exports, "MastraModelGateway", {
  enumerable: true,
  get: function () { return chunkJBYS3POB_cjs.MastraModelGateway; }
});
